---
title: 'Lea Hip & Tibia BRBseq July 2024 '
output:
  html_document:
    df_print: paged
---

```{r}
setwd("/home/mbotos/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/")
```



Path to the data: ***Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq***


# --- Import libraries
```{r}
#library(BiocManager)
rm(list = ls(all.names = TRUE))
gc()
#.libPaths("/home/mbotos/R/x86_64-pc-linux-gnu-library/4.2/")
.libPaths("/home/mbotos/CLUSTER/RLibs")
#install.packages("BiocManager")
library("BiocManager")
#BiocManager::install(c("ggplot2","edgeR","DESeq2","dorothea"),force=TRUE,ask = TRUE)
library(dorothea)

#update.packages()


library(dplyr)
library(tidyr)
library(edgeR)
library(ggplot2)
```



# --- Read tables
```{r}
matrix_counts_tibia <- read.table(file ="/home/mbotos/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/brbseq-nf-pipeline/results/AMP0211/count_matrix/L270524LF01_01.read.counts.sampleIDs.detailed.txt",sep = "\t",header = TRUE)
head(matrix_counts_tibia)
tail(matrix_counts_tibia)
dim(matrix_counts_tibia)
head(matrix_counts_tibia)
```
```{r}
matrix_counts_hip <- read.table(file ="/home/mbotos/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/brbseq-nf-pipeline/results/AMP0211/count_matrix/L270524LF02_01.read.counts.sampleIDs.detailed.txt",sep = "\t",header = TRUE)
head(matrix_counts_hip)
tail(matrix_counts_hip)
dim(matrix_counts_hip)
head(matrix_counts_hip)
```


```{r}
ncol(matrix_counts_tibia)
colnames(matrix_counts_tibia)
```

```{r}
ncol(matrix_counts_hip)
colnames(matrix_counts_hip)
```


```{r}
# Remove the last column
# New version do not need to be removed, Uknown columns is not there anymore...
#the Gene_id or Gene_name which are in the position 1 and 2

#matrix_counts <- matrix_counts[,-ncol(matrix_counts)]

# Remove the last 5 rows
matrix_counts_tibia <- matrix_counts_tibia[1:(nrow(matrix_counts_tibia)-5),]
head(matrix_counts_tibia)
tail(matrix_counts_tibia)
dim(matrix_counts_tibia)
colnames(matrix_counts_tibia)[3:ncol(matrix_counts_tibia)]
```
```{r}
# Remove the last 5 rows
matrix_counts_hip <- matrix_counts_hip[1:(nrow(matrix_counts_hip)-5),]
head(matrix_counts_hip)
tail(matrix_counts_hip)
dim(matrix_counts_hip)
colnames(matrix_counts_hip)[3:ncol(matrix_counts_hip)]
```


# --- Add metadata table from the xlsx file
```{r}
#install.packages("readxl")
md <- readxl::read_xlsx(path ="/home/mbotos/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/metadata/sample_info_minimal_SIT.xlsx")
md$sampleID
```

```{r}

# Use mutate to format the sample IDs
md_df <- md %>% 
  dplyr::mutate(sampleID = gsub(" ", ".", sampleID)) %>%
  dplyr::mutate(sampleID = paste0("X", sampleID))

# Display the data frame with formatted sample IDs
print(md_df)
```



```{r}
md
md_df

```


# Generate more columns for the data to plot
```{r}
md_df$sampleID
# Split the sampleID column
md_df_split <- strsplit(as.character(md_df$sampleID), "_", fixed = TRUE)

# Create new columns
md_df$sample_number <- sapply(md_df_split, function(x) x[1])
md_df$tissue <- sapply(md_df_split, function(x) x[2])
#md_df$health_status <- sapply(md_df_split, function(x) paste(x[3], x[4], sep = "_"))
md_df$health_status <- sapply(md_df_split, function(x) paste(x[3], sep = "_"))
md_df$tissue_celltype <- sapply(md_df_split, function(x) paste(x[4], sep = "_"))
md_df$tissue_celltype_lowercase <- tolower(md_df$tissue_celltype)
md_df$tissue_lowercase <- tolower(md_df$tissue)
md_df$full_name <- md_df$sampleID


# View the dataframe
md_df

```

# Plot number of tibia vs hip
```{r}
# Summarize the data
summary_df <- md_df %>%
  group_by(tissue_lowercase) %>%
  summarise(n = n()) %>%
  ungroup()

# Plot with individual bar colors and labels
ggplot(data = summary_df, aes(x = tissue_lowercase, y = n, fill = tissue_lowercase)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = n), vjust = -0.5, size = 3.5) +
  labs(
    title = "Number of Samples per Tissue",
    x = "Tissue",
    y = "Number of Samples"
  ) +
    theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "none",  # hide legend if each bar is already labeled
    panel.grid.major = element_blank(),  # Remove major grid lines
    panel.grid.minor = element_blank(),  # Remove minor grid lines
    panel.background = element_blank(),   # Remove background
    axis.line = element_line(colour = "black")  # Add axis lines
    
  ) +
  scale_y_continuous(breaks = seq(0, max(summary_df$n), by = 10))

```


# Add gender, age and BMI to metadata.
```{r}
agbmi_df <- readxl::read_xlsx(path ="/home/mbotos/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples_2/metadata/age-gender-BRBseq_2.xlsx")
agbmi_df

# Add X infront of the sample ID and close non sclerotic with non.sclerotic.

agbmi_df <- agbmi_df |>  
  dplyr::mutate(`Sample ID` = gsub(" ", ".", `Sample ID`)) |> 
  dplyr::mutate(`Sample ID` = paste0("X", `Sample ID`))
agbmi_df

md_df <- merge(x = md_df,y = agbmi_df,by.x = "sampleID",by.y = "Sample ID")

```







# Orientation of Data
## Transposed Data (samples as rows, genes as columns)

m <- prcomp(t(matrix_counts_hip[,3:ncol(matrix_counts_hip)]), scale. = TRUE, center = TRUE)
Input Matrix: After transposing, each row represents a sample, and each column represents a gene.
Interpretation: PCA will identify principal components based on the variance among samples across different genes.
Results: The principal components will describe the variation between the samples. Each principal component (PC) is a linear combination of the gene expression values.


## Non-Transposed Data (genes as rows, samples as columns)

m <- prcomp(matrix_counts_hip[,3:ncol(matrix_counts_hip)], scale. = TRUE, center = TRUE)
Input Matrix: Without transposing, each row represents a gene, and each column represents a sample.
Interpretation: PCA will identify principal components based on the variance among genes across different samples.
Results: The principal components will describe the variation between the genes. Each principal component (PC) is a linear combination of the sample values.


# Choosing the Right Approach
The choice of whether to transpose the data or not depends on what you want to analyze:

## Analyzing Variation Between Samples: If your goal is to analyze how samples (conditions) vary based on gene expression profiles, you should transpose the data so that each row represents a sample. This will provide principal components that describe the differences between samples.
m <- prcomp(t(matrix_counts_hip[,3:ncol(matrix_counts_hip)]), scale. = TRUE, center = TRUE)


## Analyzing Variation Between Genes: If your goal is to analyze how genes vary across different samples, you should not transpose the data, keeping genes as rows and samples as columns. This will provide principal components that describe the differences between genes.
m <- prcomp(matrix_counts_hip[,3:ncol(matrix_counts_hip)], scale. = TRUE, center = TRUE)



# Clean rows that have gene expression for all the samples < 3 counts.
```{r}
# If no results of DEG obtained, we reduce stringency, so lower the number.

ncol(matrix_counts_hip)
dim(matrix_counts_hip[,3:ncol(matrix_counts_hip)][rowSums(matrix_counts_hip[,3:ncol(matrix_counts_hip)]) >= 30,])
dim(matrix_counts_tibia[,3:ncol(matrix_counts_tibia)][rowSums(matrix_counts_tibia[,3:ncol(matrix_counts_tibia)]) >= 30,])

```


```{r}
# row_sums <- rowSums(matrix_counts_hip[, 3:ncol(matrix_counts_hip)])
# matrix_counts_hip_filtered <- matrix_counts_hip[row_sums >= 30,]
# matrix_counts_hip_filtered
```




#Function for PCA, removing empty genes counts and empty columns if trnasposed and clean duplicated
```{r}
run_pca_structuring <- function(matrix_data,transpose=FALSE,filter_empty_samples_counts=30){
  
  if (!is.data.frame(matrix_data)) {
  stop("Input matrix_data must be a data frame, not a matrix or list.")
  }

  data_matrix_raw <- matrix_data
  #check duplications, they are quite awkward gene namse =()
  #data_matrix_raw <- data_matrix_raw[!duplicated(data_matrix_raw$Gene_name), ]
  data_matrix_raw <- data_matrix_raw[!duplicated(data_matrix_raw$Gene_name) & data_matrix_raw$Gene_name != "", ]

  rownames(data_matrix_raw) <- data_matrix_raw$Gene_name
  
  data_matrix_raw <- data_matrix_raw[,3:ncol(data_matrix_raw)]
  row_sums = rowSums(data_matrix_raw)
  data_matrix_filtered = data_matrix_raw[row_sums >= filter_empty_samples_counts,]
  #data_matrix <- data_matrix_filtered
  # 
  # # Remove constant/zero columns
  # non_constant_columns <- apply(data_matrix_filtered, 2, function(x) sd(x) != 0)
  # data_matrix_filtered <- data_matrix_filtered[, non_constant_columns]
  
  # Check all are numeric
  if (!all(sapply(data_matrix_raw, is.numeric))) {
    stop("Not all columns in data_matrix_raw are numeric. PCA will fail.")
  }
  
  #Run PCA
  # Check transpose
  if (transpose){
    m <- prcomp(t(x = data_matrix_filtered),scale. = TRUE,center = TRUE)
  } else {
    m <- prcomp(x = data_matrix_filtered, scale. = TRUE,center = TRUE)
  }
  #m <- prcomp(data_matrix,scale. = TRUE,center = TRUE)
  
  #calculat the PC "%"'s
  percentVar_obj <- m$sdev^2/sum(m$sdev^2)
  
  #Save as a df
  m <- as.data.frame(m[2]$rotation)
  m <- merge(x = m,y = md_df,by.x="row.names",by.y="full_name")
  
  return(list(
    matrix_raw=data_matrix_filtered,
    matrix_pca=m,
    percentVar_obj=percentVar_obj))
  
}

```

# Check the duplicated such as SNORA,RGS5, SIGLEC5,SPATA13,U3,U2,U8,SNORD,YRNA..

```{r}
#Genes variability, typically biology difference across samples is stronger.. so it is also difficult to catch...
m_pca_tibia <- run_pca_structuring(matrix_data = matrix_counts_tibia,transpose = FALSE,filter_empty_samples_counts = 30)
m_pca_hip <- run_pca_structuring(matrix_data = matrix_counts_hip,transpose = FALSE,filter_empty_samples_counts = 30)

```

### If this crash, make sure that the matrix is a matrix.
```{r}
# # If the list is a list of equal-length numeric vectors:
# data_matrix_fixed <- as.data.frame(data_matrix_filtered)
# 
# # Or try to coerce directly to a matrix:
# data_matrix_fixed <- as.matrix(data_matrix_filtered)
# 
# str(data_matrix_filtered)
# is.matrix(data_matrix_filtered)

```


# Plot the PCA function
```{r}
require("ggplot2")
#shapes
plot_the_pca <- function(pca_df,colors,pca,pcb,title,subtitle,colors_selection,shapes,shapes_selection,alphas=0.6,text_column="Row.names",percentVar_obj) {
  
  n <- ggplot(data = pca_df,aes(x = .data[[paste0("PC",substr(pca,start = 3,stop = 3))]],
                                y = .data[[paste0("PC",substr(pcb,start = 3,stop = 3))]],
                                color = .data[[colors]],
                                shape = .data[[shapes]]))
  # n <- n + geom_point(size=6)
  n <- n + geom_jitter(position = position_jitter(width = 0.000000001,height = 0.000000001,seed = 123456),size=6,alpha=alphas)
  n <- n + labs(title=paste0("PCA ",title),
                #color = paste0(colors),
                #shape = paste0(shapes),
                subtitle = paste0(subtitle,"")) +
    xlab(paste0("PC",substr(paste0("PC",substr(pca,start = 3,stop = 3)),start = 3,stop = 3),": ",
                round(percentVar_obj[as.numeric(substr(paste0("PC",substr(pca,start = 3,stop = 3)),start = 3,stop = 3))] * 100,digits = 2), "% variance")) +
    ylab(paste0("PC",substr(paste0("PC",substr(pcb,start = 3,stop = 3)),start = 3,stop = 3),": ",
                round(percentVar_obj[as.numeric(substr(paste0("PC",substr(pcb,start = 3,stop = 3)),start = 3,stop = 3))] * 100,digits = 2), "% variance"))
  n <- n + theme(panel.grid = element_blank(),
                 panel.border = element_rect(fill = "transparent"),
                 panel.background = element_rect(fill = "transparent"),
                 legend.title = element_text(size=15),
                 legend.text = element_text(size=12))
  n <- n + geom_text(aes(label = .data[[text_column]]),color = "black", size = 2)

  # n <- n + scale_color_manual(values = colors_selection)
  # n <- n + scale_shape_manual(values = shapes_selection)
  n
  
}
```


# --- Raw

# Tibia
```{r,fig.width=6,fig.height=3}
### Need to be run the run_pca_structuring before the plot the pca otherwise the objects mess up
# Generate the columns in the md_df
plot_the_pca(pca_df = m_pca_tibia$matrix_pca,
             percentVar_obj = m_pca_tibia$percentVar_obj,
             colors =  "tissue_celltype_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias SVF and Adipocytes",
             text_column = "sample_number")

```



# Hip
```{r,fig.width=6,fig.height=3}
### Need to be run the run_pca_structuring before the plot the pca otherwise the objects mess up
# Generate the columns in the md_df
plot_the_pca(pca_df = m_pca_hip$matrix_pca,
             percentVar_obj = m_pca_hip$percentVar_obj,
             colors =  "tissue_celltype_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Hips SVF and Adipocytes",
             text_column = "sample_number")
```


```{r}
m_pca_tibia_t <- run_pca_structuring(matrix_data = matrix_counts_tibia,transpose = TRUE,filter_empty_samples_counts = 30)
m_pca_hip_t <- run_pca_structuring(matrix_data = matrix_counts_hip,transpose = TRUE,filter_empty_samples_counts = 30)

```

# Tibia
```{r,fig.width=6,fig.height=3}
### Need to be run the run_pca_structuring before the plot the pca otherwise the objects mess up
# Generate the columns in the md_df
plot_the_pca(pca_df = m_pca_tibia_t$matrix_pca,
             percentVar_obj = m_pca_tibia_t$percentVar_obj,
             colors =  "tissue_celltype_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias SVF and Adipocyte Transposed",
             text_column = "sample_number")

```
# Hip
```{r,fig.width=6,fig.height=3}
### Need to be run the run_pca_structuring before the plot the pca otherwise the objects mess up
# Generate the columns in the md_df
plot_the_pca(pca_df = m_pca_hip_t$matrix_pca,
             percentVar_obj = m_pca_hip_t$percentVar_obj,
             colors =  "tissue_celltype_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",
             subtitle = "Hips SVF and Adipocyte Transposed",
             text_column = "sample_number")

```


#############################
#     Tibia Adipocytes      #
#############################
```{r,fig.width=6,fig.height=3}
m_pca_tibia_transposed_adipocytes <- run_pca_structuring(matrix_data = matrix_counts_tibia |> dplyr::select("Gene_id","Gene_name",ends_with("adipocyte")),transpose = TRUE,filter_empty_samples_counts = 30)

plot_the_pca(pca_df = m_pca_tibia_transposed_adipocytes$matrix_pca,
             percentVar_obj = m_pca_tibia_transposed_adipocytes$percentVar_obj,
             colors =  "tissue_celltype_lowercase",
             shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias Adipocytes Transposed",
             text_column = "sample_number")



# Issued samples with low number of genes per sample from HTML report
#107 tibia non sclerotic adipocyte
#107 tibia sclerotic adipocyte
##99 tibia sclerotic adipocyte


#80 tibia non sclerotic pellet
#163 tibia sclerotic pellet
```

#############################
#     Tibia SVF/Pellet      #
#############################
```{r,fig.width=6,fig.height=3}
m_pca_tibia_transposed_pellet <- run_pca_structuring(matrix_data = matrix_counts_tibia |> dplyr::select("Gene_id","Gene_name",ends_with("pellet")),transpose = TRUE,filter_empty_samples_counts = 30)


plot_the_pca(pca_df = m_pca_tibia_transposed_pellet$matrix_pca,
             percentVar_obj = m_pca_tibia_transposed_pellet$percentVar_obj,
             colors =  "tissue_celltype_lowercase",
             shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias Pellet Transposed",
             text_column = "sample_number")

# Issued samples with low number of genes per sample from HTML report

#80 tibia non sclerotic pellet
#163 tibia sclerotic pellet
```

```{r}

```
#############################
#     Hip Adipocytes        #
#############################
```{r,fig.width=6,fig.height=3}
m_pca_hip_transposed_adipocytes <- run_pca_structuring(matrix_data = matrix_counts_hip |> dplyr::select("Gene_id","Gene_name",ends_with("adipocyte")),transpose = TRUE,filter_empty_samples_counts = 30)

plot_the_pca(pca_df = m_pca_hip_transposed_adipocytes$matrix_pca,
             percentVar_obj = m_pca_hip_transposed_adipocytes$percentVar_obj,
             colors =  "tissue_celltype_lowercase",
             shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Hips Adipocytes Transposed",
             text_column = "sample_number")



# Issued samples with low number of genes per sample from HTML report, expected to be also associated with low number of sequencing read per sample...
#74 hip non sclerotic adipocyte
#79 hip non sclerotic adipocyte
#150 hip non sclerotic adipocyte

#74 hip sclerotic adipocyte
#79 hip sclerotic adipocyte

#78 hip non sclerotic adipocyte
#78 hip sclerotic adipocyte

```


#############################
#     Hip SVF/Pellet      #
#############################
```{r,fig.width=6,fig.height=3}
m_pca_hip_transposed_pellet <- run_pca_structuring(matrix_data = matrix_counts_hip |> dplyr::select("Gene_id","Gene_name",ends_with("pellet")),transpose = TRUE,filter_empty_samples_counts = 30)


plot_the_pca(pca_df = m_pca_hip_transposed_pellet$matrix_pca,
             percentVar_obj = m_pca_hip_transposed_pellet$percentVar_obj,
             colors =  "tissue_celltype_lowercase",
             shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Hips Pellet Transposed",
             text_column = "sample_number")

# Issued samples with low number of genes per sample from HTML report

 
#74 Hip non sclerotic pellet
#74 Hip sclerotic pellet
#78 hip non.sclerotic pellet
#78 hip sclerotic pellet
```

```{r}
md_df
```


```{r}
plot_the_pca(pca_df = m_pca_hip_transposed_pellet$matrix_pca,
             percentVar_obj = m_pca_hip_transposed_pellet$percentVar_obj,
             colors =  "tissue_celltype_lowercase",
             shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Hips Pellet Transposed",
             text_column = "Age")
```

```{r}
plot_the_pca(pca_df = m_pca_hip_transposed_pellet$matrix_pca,
             percentVar_obj = m_pca_hip_transposed_pellet$percentVar_obj,
             colors =  "Gender",
             shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Hips Pellet Transposed",
             text_column = "Age")
```
##########################
# Test of QC for samples #
##########################


## Remove if are away for complexity or number of genes.
Genes here have been filtered for under 30 counts.. in the matrix_raw counts

```{r}
m_pca_tibia$matrix_raw
```
### Plot the number of reads per sample from raw matrix of tibia.
```{r,fig.width=25,fig.height=9}
#matrix_counts_tibia
#m_pca_tibia$matrix_raw

sums_mm = colSums(m_pca_tibia$matrix_raw[3:ncol(m_pca_tibia$matrix_raw)])
sums_mm <- data.frame(Column = colnames(m_pca_tibia$matrix_raw[3:ncol(m_pca_tibia$matrix_raw)]), Sum = sums_mm)
sums_mm$Column <-tolower(sums_mm$Column)



# Categorize the groups based on the Column names
sums_mm$Group <- ifelse(grepl("tibia_non.sclerotic_adipocyte", sums_mm$Column), "TNSA",
                   ifelse(grepl("tibia_sclerotic_adipocyte", sums_mm$Column), "TSA",
                          ifelse(grepl("tibia_non.sclerotic_pellet", sums_mm$Column), "TNSP", "TSP")))

sums_mm
# Create a custom color palette
color_palette <- c("TNSA" = "#00296b", "TSA" = "#1e91d0", "TNSP" = "#faa819","TSP"="#f37520")
```


```{r,fig.width=29,fig.height=14}
library(ggplot2)
mean_sum <- mean(sums_mm$Sum)
quantiles <- quantile(sums_mm$Sum, probs = c(0.25, 0.75,0.90))

p <- ggplot(sums_mm, aes(x = reorder(Column, Sum), y = Sum, fill = Group)) +
#ggplot(sums_mm, aes(x = Column, y = Sum, fill = Group)) +
  

  geom_bar(stat = "identity") +
  labs(title = "Counts Sums",
       x = "Samples",
       y = "Sum of all counts") +
  scale_fill_manual(values = color_palette) +
  geom_hline(aes(yintercept = mean_sum, color = "Mean"), linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = quantiles[1], color = "25th Percentile"), linetype = "dotted", linewidth = 1) +
  geom_hline(aes(yintercept = quantiles[2], color = "75th Percentile"), linetype = "dotted", linewidth = 1) +
  geom_hline(aes(yintercept = quantiles[3], color = "90th Percentile"), linetype = "dotted", linewidth = 1) +
  scale_color_manual(values = c("red", "blue", "green","#903498"),
                     breaks = c("Mean", "25th Percentile", "75th Percentile", "90th Percentile"),
                     name = "Statistics") +  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        axis.text.y = element_text(size = 12),
        panel.background = element_blank(),  # Remove panel background
        panel.grid.major = element_blank(),  # Remove major grid lines
        panel.grid.minor = element_blank(),  # Remove minor grid lines
        )  # Rotate x-axis labels for better visibility

pp <- plotly::ggplotly(p)
pp
```
# TIBIA Samples out

#X163   TSP
#X99    TSA
#X107   TNSA TSA TNSP
#X80    TNSP TNSA TSP
#X105   TSP TNSP
#X148   TSP
#X104   TSA TSP
#X164   TSA 
#X109   TNSP
# we filter out around same proportions if 25% in TIBIA is around 700k we filtered 407265 and in HIP 25% is 1M in Hip, we filterd 791K


```{r}
sums_mm |> dplyr::filter(Sum < 407265)
sums_mm_tibia <- sums_mm |> dplyr::filter(!Sum < 407265)
sums_mm_tibia$Samples <- rownames(sums_mm_tibia)
sums_mm_tibia
```

# Extra filtered from the group out
```{r}
#107 All
#80 All
sums_mm_tibia <- sums_mm_tibia |> dplyr::filter(!Samples %in% c("X107_tibia_sclerotic_pellet",
                                                                "X80_tibia_sclerotic_adipocyte"))
sums_mm_tibia

```



# HIP

```{r}
m_pca_hip$matrix_raw
```
### Plot the number of reads per sample from raw matrix of tibia.
```{r,fig.width=25,fig.height=9}
#matrix_counts_hip
#m_pca_hip$matrix_raw
sums_mm = colSums(m_pca_hip$matrix_raw[3:ncol(m_pca_hip$matrix_raw)])
sums_mm <- data.frame(Column = colnames(m_pca_hip$matrix_raw[3:ncol(m_pca_hip$matrix_raw)]), Sum = sums_mm)


sums_mm$Column <-tolower(sums_mm$Column)


# Categorize the groups based on the Column names
sums_mm$Group <- ifelse(grepl("hip_non.sclerotic_adipocyte", sums_mm$Column), "HNSA",
                   ifelse(grepl("hip_sclerotic_adipocyte", sums_mm$Column), "HSA",
                          ifelse(grepl("hip_non.sclerotic_pellet", sums_mm$Column), "HNSP", "HSP")))

sums_mm
# Create a custom color palette
color_palette <- c("HNSA" = "#00296b", "HSA" = "#1e91d0", "HNSP" = "#faa819","HSP"="#f37520")
```


```{r,fig.width=25,fig.height=9}
library(ggplot2)
c <- mean(sums_mm$Sum)
quantiles <- quantile(sums_mm$Sum, probs = c(0.25, 0.75,0.90))

p <- ggplot(sums_mm, aes(x = reorder(Column, Sum), y = Sum, fill = Group)) +
#ggplot(sums_mm, aes(x = Column, y = Sum, fill = Group)) +
  

  geom_bar(stat = "identity") +
  labs(title = "Counts Sums",
       x = "Samples",
       y = "Sum of all counts") +
  scale_fill_manual(values = color_palette) +
  geom_hline(aes(yintercept = mean_sum, color = "Mean"), linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = quantiles[1], color = "25th Percentile"), linetype = "dotted", size = 1) +
  geom_hline(aes(yintercept = quantiles[2], color = "75th Percentile"), linetype = "dotted", size = 1) +
  geom_hline(aes(yintercept = quantiles[3], color = "90th Percentile"), linetype = "dotted", size = 1) +
  scale_color_manual(values = c("red", "blue", "green","#903498"),
                     breaks = c("Mean", "25th Percentile", "75th Percentile", "90th Percentile"),
                     name = "Statistics") +  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        axis.text.y = element_text(size = 12),
        panel.background = element_blank(),  # Remove panel background
        panel.grid.major = element_blank(),  # Remove major grid lines
        panel.grid.minor = element_blank(),  # Remove minor grid lines
        )  # Rotate x-axis labels for better visibility

plotly::ggplotly(p)
```
# HIP Samples out
#####X78 - 4
#####X74 -4
#####X79 -2 HNSA and HSA
#####X150 HNSA
#####X154 HNSP
# Removed all the samples with less than 791286 counts in Hip.

# Filter Samples
```{r}
sums_mm_hip <- sums_mm |> dplyr::filter(!Sum < 791286)
sums_mm |> dplyr::filter(Sum < 791286)
sums_mm_hip$Samples <- rownames(sums_mm_hip)
sums_mm_hip

```


# Comparisons
### Use the samples cleaned objects

```{r}
sums_mm_tibia
sums_mm_hip 
```

# filter out the samples we removed before
```{r}
sums_mm_tibia
sums_mm_hip
#matrix_counts_tibia
#matrix_counts_hip

m_pca_tibia$matrix_raw_filtered_samples_for_DEG <- m_pca_tibia$matrix_raw[,rownames(sums_mm_tibia)]
m_pca_hip$matrix_raw_filtered_samples_for_DEG <- m_pca_hip$matrix_raw[,rownames(sums_mm_hip)]
#using dplyr
#m_pca_tibia$matrix_raw_filtered_samples_for_DEG <- m_pca_tibia$matrix_raw |>   select(all_of(rownames(sums_mm_tibia)))
```


```{r}
# Merge back the metadata and the selected samples 
md_df_tibia <- merge(x=md_df,y = sums_mm_tibia,by.x="sampleID",by.y="Samples")
md_df_hip <- merge(x=md_df,y = sums_mm_hip,by.x="sampleID",by.y="Samples")
md_df_tibia$sampleID[grepl(pattern = "X80",x = md_df_tibia$sampleID)]
md_df_hip$sampleID[grepl(pattern = "X74",x = md_df_hip$sampleID)]
```


#############
### EDGER ###
#############

### LRT: Suitable for complex designs but may not fully account for variability, potentially leading to inflated false positives.
### QLF: Provides robust results by modeling extra variability through quasi-likelihood methods and shrinkage estimators. It yields more accurate p-values and better control over false positives.

# --- Normalize
```{r,fig.width=12,fig.height=6}
#############
#   edgeR   #
#############
library(data.table)
library(edgeR)
#Run edgeR on the COMBATseq batch corrected data.
message("Builduing the DEG data list for edgeR.")

#.1- Build DGE list and normalize
y = DGEList(counts = m_pca_tibia$matrix_raw_filtered_samples_for_DEG, group = md_df_tibia$Group, genes = m_pca_tibia$matrix_raw[,2])
#perform TMM normalization
message("Normalizing using TMM.\n")
y_norm <- calcNormFactors(object = y,method = "TMM")
y_normalized_corrected <- cpm(y = y_norm,normalized.lib.sizes = TRUE)
rownames(y_normalized_corrected) <- y$genes$genes

#y_normalized_corrected_pcas <- prcomp(log2(y_normalized_corrected + 1))
y_normalized_corrected_pcas <- prcomp(y_normalized_corrected,center = TRUE,scale. = TRUE)


#calculat the PC "%"'s
percentVar_obj <- y_normalized_corrected_pcas$sdev^2/sum(y_normalized_corrected_pcas$sdev^2)
#Save as a df
y_normalized_corrected_pcas_df <- as.data.frame(y_normalized_corrected_pcas[2]$rotation)
y_normalized_corrected_pcas_df <- merge(y_normalized_corrected_pcas_df,md_df_tibia, by.x = "row.names", by.y = "sampleID")
# y_normalized_corrected_pcas_df[,c(1,18)]
message("Done")
message("Plot ready.")


# y_normalized_corrected_pcas_df <- y_normalized_corrected_pcas_df |> dplyr::mutate(Sample2 = sub("^(.*?)\\..*", "\\1", Sample))
# y_normalized_corrected_pcas_df <- y_normalized_corrected_pcas_df |> dplyr::mutate(Sample3 = sub(".*?\\.(.*)", "\\1", Sample))


#colnames(md_df_tibia)
#md_df_tibia$Age <- as.factor(md_df_tibia$Age)
for (column_to_plot in c("Tissue/cell type","Age","health_status","Group","Plate.ID")) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
                     percentVar_obj = percentVar_obj,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias (Normalized TMM)",subtitle = "TSA TSP TNSA TNSP",
                     text_column = "sample_number"))
  }

```


# EDGER Normalization function

```{r}
library(edgeR)
library(dplyr)

normalize_and_pca <- function(count_matrix, gene_info, metadata,transpose=FALSE) {
  # Check if inputs are data frames or matrices
  if (!is.data.frame(count_matrix) && !is.matrix(count_matrix)) {
    stop("count_matrix must be a data frame or matrix.")
  # }
  # if (!is.factor(group)) {
  #   stop("group must be a factor.")
  }
  if (!is.vector(gene_info) && !is.list(gene_info)) {
    stop("gene_info must be a vector or list")
  }
  if (!is.data.frame(metadata)) {
    stop("metadata must be a data frame.")
  }
  
  # Build DGE list and normalize
  message("Building DGEList and normalizing using TMM.\n")
  y <- DGEList(counts = count_matrix, group = metadata$Group, genes = gene_info)
  y_norm <- calcNormFactors(y, method = "TMM")
  y_normalized_corrected <- cpm(y_norm, normalized.lib.sizes = TRUE)
  rownames(y_normalized_corrected) <- y$genes$genes
  
  # Perform PCA
  message("Performing PCA.\n")
    # Check transpose
  if (transpose){
    y_normalized_corrected_pcas <- prcomp(t(y_normalized_corrected), center = TRUE, scale. = TRUE)
  } else {
    y_normalized_corrected_pcas <- prcomp(y_normalized_corrected, center = TRUE, scale. = TRUE)
  }
  #y_normalized_corrected_pcas <- prcomp(y_normalized_corrected, center = TRUE, scale. = TRUE)
  
  
    
  # Calculate the percentage of variance for each principal component
  percentVar_obj <- y_normalized_corrected_pcas$sdev^2 / sum(y_normalized_corrected_pcas$sdev^2)
  
  # Save PCA results as a data frame
  y_normalized_corrected_pcas_df <- as.data.frame(y_normalized_corrected_pcas[2]$rotation)
  
  # Merge PCA results with metadata
  y_normalized_corrected_pcas_df <- merge(y_normalized_corrected_pcas_df, metadata, by.x = "row.names", by.y = "sampleID")
  
  # # Optionally mutate the sample names
  # y_normalized_corrected_pcas_df <- y_normalized_corrected_pcas_df %>%
  #   mutate(Sample2 = sub("^(.*?)\\..*", "\\1", Row.names),
  #          Sample3 = sub(".*?\\.(.*)", "\\1", Row.names))
  # 
  message("Done")
  message("ready to Plot ")
  
  return(list(
    pca_results = y_normalized_corrected_pcas_df,
    percent_variance = percentVar_obj
  ))
}

# Example usage:

results <- normalize_and_pca(
  count_matrix = m_pca_tibia$matrix_raw_filtered_samples_for_DEG,
  gene_info = rownames(m_pca_tibia$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_tibia
)

#md_df_tibia$Age <- as.factor(md_df_tibia$Age)
for (column_to_plot in c("Tissue/cell type","Age","health_status","Group", "Plate.ID")) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias (Normalized TMM)",subtitle = "TSA TSP TNSA TNSP",
                     text_column = "sample_number"))
}
```

# Tibia only pellet
```{r warning=FALSE}
md_df_tibia |>
  dplyr::filter(tissue_celltype_lowercase == "pellet") |>
  dplyr::pull(sampleID)
# colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet"))


results <- normalize_and_pca(
  count_matrix = m_pca_tibia$matrix_raw_filtered_samples_for_DEG[,md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet") |> dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_tibia$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet",),
  transpose = FALSE
)


for (column_to_plot in c(colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias (Normalized TMM)",subtitle = "TSP TNSP",
                     text_column = "sample_number"))
}

```
#Tibia only adipocyte
```{r warning=FALSE}
md_df_tibia |>
  dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
  dplyr::pull(sampleID)
# colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet"))



results <- normalize_and_pca(
  count_matrix = m_pca_tibia$matrix_raw_filtered_samples_for_DEG[,md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "adipocyte") |> dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_tibia$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "adipocyte"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "adipocyte")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias (Normalized TMM)",subtitle = "TSA TNSA",
                     text_column = "sample_number"))
}
```



# Lea
# Biological observation
#########################################################################################
# Leas observation, might be the super sclerotic cluster female based?
# Might create a super slerotic samples and compare them to non-sclerotic or mixed? (Wait to confirm with slides)
#########################################################################################



###########################
#      HIP                #
###########################
# Normalized PCA
```{r}
results <- normalize_and_pca(
  count_matrix = m_pca_hip$matrix_raw_filtered_samples_for_DEG,
  gene_info = rownames(m_pca_hip$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_hip,
  transpose = FALSE)

#md_df_tibia$Age <- as.factor(md_df_tibia$Age)
for (column_to_plot in c("Tissue/cell type","Age","health_status","Group", "Plate.ID")) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Hips (Normalized TMM)",subtitle = "HSA HSP HNSA HNSP",
                     text_column = "sample_number"))
}

```



#Hip only adipocyte
```{r warning=FALSE}
md_df_hip |>
  dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
  dplyr::pull(sampleID)
# colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet"))



results <- normalize_and_pca(
  count_matrix = m_pca_hip$matrix_raw_filtered_samples_for_DEG[,md_df_hip |>
                                                                 dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                 dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_hip$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "adipocyte"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "adipocyte")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Hips (Normalized TMM)",subtitle = "HSA HNSA",
                     text_column = "sample_number"))
}
```
#Hip only pellet
```{r warning=FALSE}
md_df_hip |>
  dplyr::filter(tissue_celltype_lowercase == "pellet") |>
  dplyr::pull(sampleID)
# colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet"))



results <- normalize_and_pca(
  count_matrix = m_pca_hip$matrix_raw_filtered_samples_for_DEG[,md_df_hip |>
                                                                 dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                 dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_hip$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "pellet"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "pellet")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Hips (Normalized TMM)",subtitle = "HSP HNSP",
                     text_column = "sample_number"))
}
```

# Add up and check differences only between sclerotic and non sclerotic
# Generate metadata
```{r}
md_df_tibia$barcode_tissue <- paste0(md_df_tibia$barcode,"_tibia")
md_df_tibia$Plate.ID_tissue <- paste0(md_df_tibia$Plate.ID,"_tibia")
md_df_hip$barcode_tissue <- paste0(md_df_hip$barcode,"_hip")
md_df_hip$Plate.ID_tissue <- paste0(md_df_hip$Plate.ID,"_hip")

md_df_bone <- rbind(md_df_tibia,md_df_hip)

md_df_bone |>
  dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
  dplyr::pull(sampleID)
# colnames(md_df_tibia |> dplyr::filter(tissue_celltype_lowercase == "pellet"))
```

#Generate matrix for PCA of samples together.
```{r}
# m_pca_tibia <- run_pca_structuring(matrix_data = matrix_counts_tibia,transpose = FALSE,filter_empty_samples_counts = 30)
# m_pca_hip <- run_pca_structuring(matrix_data = matrix_counts_hip,transpose = FALSE,filter_empty_samples_counts = 30)


path="~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/"
# Save genes unique to TIBIA
length(setdiff(rownames(m_pca_tibia$matrix_raw_filtered_samples_for_DEG),rownames(m_pca_hip$matrix_raw_filtered_samples_for_DEG)))
write.table(x = setdiff(rownames(m_pca_tibia$matrix_raw_filtered_samples_for_DEG),rownames(m_pca_hip$matrix_raw_filtered_samples_for_DEG)),
            file = paste0(path,"unique_genes_of_TIBIA.txt"),sep = "\t", row.names = FALSE, col.names = FALSE)

# Save genes unique to HIP
length(setdiff(rownames(m_pca_hip$matrix_raw_filtered_samples_for_DEG),rownames(m_pca_tibia$matrix_raw_filtered_samples_for_DEG)))
write.table(x = setdiff(rownames(m_pca_hip$matrix_raw_filtered_samples_for_DEG),rownames(m_pca_tibia$matrix_raw_filtered_samples_for_DEG)),
            file = paste0(path,"unique_genes_of_HIP.txt"),sep = "\t", row.names = FALSE, col.names = FALSE)


# Find common genes between TIBIA and HIP datasets
common_genes <- intersect(rownames(m_pca_tibia$matrix_raw_filtered_samples_for_DEG), rownames(m_pca_hip$matrix_raw_filtered_samples_for_DEG))

# Save the list of common genes to a file
write.table(x = common_genes, file = paste0(path, "common_genes_to_TIBIA_and_HIP.txt"), sep = "\t", row.names = FALSE, col.names = FALSE)

# Combine the data for the common genes from both datasets
m_samples_bone <- cbind(
  m_pca_tibia$matrix_raw_filtered_samples_for_DEG[common_genes, ],
  m_pca_hip$matrix_raw_filtered_samples_for_DEG[common_genes, ]
)
```


```{r,fig.width=12,fig.height=5}
m_samples_bone

# Calculate column sums
col_sums <- colSums(m_samples_bone)

# Create a data frame for plotting
df <- data.frame("Sample" = names(col_sums), "Sum" = col_sums)

# Merge with metadata to get group information
df <- df %>%
  left_join(md_df_bone, by = c("Sample" = "sampleID"))


# Calculate percentiles
percentiles <- quantile(df$Sum, probs = c(0.25, 0.5, 0.75, 0.90))

# Create the bar plot
ggplot(df, aes(x = reorder(Column, Sum.x), y = Sum.x, fill= Group)) +
  
  geom_bar(stat = "identity") +
  geom_hline(yintercept = percentiles, linetype = "dashed", color = "red") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(title = "Column Sums of Each Sample", x = "Sample", y = "Sum")

 
```

# STRUCTURING FUCTION 2 to avoid refiltering the samples... that are already filtered..

we look at samples similarity as is expected to be different as they are different biological humans that might dffer to much to capture only genes loadings, so we look here at PCA without t()

if there is an error due to sample variance, it means that some samples are very bad quality... either remove them or transpose to look at genes despite different samples will hdi more than genes...
```{r}
run_pca_structuring_2 <- function(matrix_data,transpose=FALSE,filter_empty_samples_counts=30){
  
  data_matrix_raw <- matrix_data
  #check duplications, they are quite awkward gene namse =()
  #data_matrix_raw <- data_matrix_raw[!duplicated(data_matrix_raw$Gene_name), ]
  #rownames(data_matrix_raw) <- data_matrix_raw$Gene_name
  
  #data_matrix_raw <- data_matrix_raw[,3:ncol(data_matrix_raw)]
  row_sums = rowSums(data_matrix_raw)
  data_matrix_filtered = data_matrix_raw[row_sums >= filter_empty_samples_counts,]
  #data_matrix <- data_matrix_filtered
  # 
  # # Remove constant/zero columns
  # non_constant_columns <- apply(data_matrix_filtered, 2, function(x) sd(x) != 0)
  # data_matrix_filtered <- data_matrix_filtered[, non_constant_columns]

  
  # Run PCA
  # Check transpose
  if (transpose) {
    m <- prcomp(t(data_matrix_filtered), scale. = TRUE, center = TRUE)
  } else {
    m <- prcomp(data_matrix_filtered, scale. = TRUE, center = TRUE)
  }

  
  #calculat the PC "%"'s
  percentVar_obj <- m$sdev^2/sum(m$sdev^2)
  
  #Save as a df
  m <- as.data.frame(m[2]$rotation)
  m <- merge(x = m,y = md_df,by.x="row.names",by.y="full_name")
  
  return(list(
    matrix_raw=data_matrix_filtered,
    matrix_pca=m,
    percentVar_obj=percentVar_obj))
  
}
```





```{r}
#m_samples_bone <- cbind(m_pca_tibia$matrix_raw,m_pca_hip$matrix_raw[3:ncol(m_pca_hip$matrix_raw)])
m_pca_samples_bone <- run_pca_structuring_2(matrix_data = m_samples_bone,transpose = FALSE,filter_empty_samples_counts = 30)
```


```{r,fig.width=12,fig.height=6}
### Need to be run the run_pca_structuring before the plot the pca otherwise the objects mess up
# Generate the columns in the md_df_bone

plot_the_pca(pca_df = m_pca_samples_bone$matrix_pca,
             percentVar_obj = m_pca_samples_bone$percentVar_obj,
             colors =  "tissue_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias and Hips",
             text_column = "tissue_celltype_lowercase")






plotly::ggplotly(
  plot_the_pca(pca_df = m_pca_samples_bone$matrix_pca,
             percentVar_obj = m_pca_samples_bone$percentVar_obj,
             colors =  "tissue_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias and Hips",
             text_column = "sample_number"))
```


```{r,fig.width=12,fig.height=6}
### Need to be run the run_pca_structuring before the plot the pca otherwise the objects mess up
# Generate the columns in the md_df_bone

plot_the_pca(pca_df = m_pca_samples_bone$matrix_pca,
             percentVar_obj = m_pca_samples_bone$percentVar_obj,
             colors =  "tissue_lowercase",shapes = "health_status",
             pca = "PC1",pcb = "PC2",
             title="PCA of RNA-seq Raw Data",subtitle = "Tibias and Hips",
             text_column = "sample_number")

```

# FILTERING SAMPLES
###################################################################################
#                                                                                 #
# filtered samples that did not pass the QC and use them for future DEG           #
#                                                                                 #
###################################################################################

```{r}
m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG <- m_pca_samples_bone$matrix_raw[,md_df_bone$sampleID]
```

# All 
```{r}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone,
  transpose = FALSE
)

for (column_to_plot in c(colnames(md_df_bone))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TSA TSP TNSA TNSP HSA HSP HNSA HNSP",
                     text_column = "sample_number"))
}

```



# Sclerotic all
```{r warning=FALSE}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   #dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::filter(health_status == "sclerotic") |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone |> dplyr::filter(health_status == "sclerotic"),
  transpose = FALSE
)

for (column_to_plot in c(colnames(md_df_bone |> dplyr::filter(health_status == "sclerotic")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TSA TSP HSA HSP",
                     text_column = "sample_number"))
}

# Adipo sclerotic 
# Adipo non sclerotic
```
# Non sclerotic all
```{r warning=FALSE}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   #dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::filter(health_status == "non.sclerotic") |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone |> dplyr::filter(health_status == "non.sclerotic"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_bone |> dplyr::filter(health_status == "non.sclerotic")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TNSA TNSP HNSA HNSP",
                     text_column = "sample_number"))
}
```




#Sclerotic only Pellets
```{r warning=FALSE}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   #dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "pellet") |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone |> dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "pellet"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_bone |> dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "pellet")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TSP HSP",
                     text_column = "sample_number"))
}
```
# Non sclerotic only pellets
```{r warning=FALSE}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   #dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "pellet") |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone |> dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "pellet"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_bone |> dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "pellet")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TNSP HNSP",
                     text_column = "sample_number"))
}
```
# Sclerotic adipocytes only
```{r warning=FALSE}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   #dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone |> dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_bone |> dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TSA HSA",
                     text_column = "sample_number"))
}
```

# Non scleoritc adipocytes only
```{r warning=FALSE}
results <- normalize_and_pca(
  count_matrix = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                   #dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                   dplyr::pull(sampleID)],
  gene_info = rownames(m_pca_samples_bone$matrix_raw),  # Ensure this is a data frame or matrix
  metadata = md_df_bone |> dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte"),
  transpose = FALSE)

for (column_to_plot in c(colnames(md_df_bone |> dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte")))) {
  #print(plotly::ggplotly(plot_the_pca(pca_df = y_normalized_corrected_pcas_df,
  print(plot_the_pca(pca_df = results$pca_results,
                     percentVar_obj = results$percent_variance,
                     colors = column_to_plot,
                     shapes = "Gender",
                     pca = "PC1", pcb = "PC2",
                     title = "Tibias and Hips (Normalized TMM)",subtitle = "TNSA HNSA",
                     text_column = "sample_number"))
}
```
# volcano function
```{r}

plot_volcano <- function(results_table, 
                         gene_column = "genes", 
                         logFC_column = "logFC", 
                         FDR_column = "FDR", 
                         genes_of_interest = NULL, 
                         output_file = paste0("volcano_plot.",format_plot),
                         #output_file = NULL,
                         format_plot = "tiff",
                         up_color = "#d00000", 
                         down_color = "#3f88c5", 
                         ns_color = "#f2e9e4", 
                         up_size = 2, 
                         down_size = 2, 
                         ns_size = 1,
                         up_alpha = 1, 
                         down_alpha = 1, 
                         ns_alpha = 0.5,
                         x_limits = c(-10, 10),
                         y_threshold = -log10(0.05),
                         x_label = "Log2 Fold Change",
                         y_label = "-Log10 Adjusted P-Value") {
  
  # Mutate the gene type
  results_table <- results_table |>
    dplyr::mutate(gene_type = dplyr::case_when(
      !!rlang::sym(logFC_column) > 1 & !!rlang::sym(FDR_column) < 0.05 ~ "Upregulated",
      !!rlang::sym(logFC_column) < -1 & !!rlang::sym(FDR_column) < 0.05 ~ "Downregulated",
      TRUE ~ "Not selected"
    )) |>
    dplyr::filter(!!rlang::sym(gene_column) != "")
  
  # Get the top up and down genes
  ints_up <- results_table |>
    dplyr::filter(!!rlang::sym(FDR_column) < 0.05 & !!rlang::sym(logFC_column) > 1) |>
    dplyr::arrange(!!rlang::sym(FDR_column)) |>
    dplyr::slice(1:10) |>
    dplyr::pull(!!rlang::sym(gene_column))
  
  ints_down <- results_table |>
    dplyr::filter(!!rlang::sym(FDR_column) < 0.05 & !!rlang::sym(logFC_column) < -1) |>
    dplyr::arrange(!!rlang::sym(FDR_column)) |>
    dplyr::slice(1:10) |>
    dplyr::pull(!!rlang::sym(gene_column))
  
  ints <- c(ints_up, ints_down)
  
  ints_genes <- results_table |>
    dplyr::filter(!!rlang::sym(gene_column) %in% ints) |>
    dplyr::distinct(!!rlang::sym(gene_column), .keep_all = TRUE)
  
  if (!is.null(genes_of_interest)) {
    to_add_genes <- results_table |>
      dplyr::filter(!!rlang::sym(gene_column) %in% genes_of_interest) |> 
      dplyr::mutate(gene_type = dplyr::case_when(
        !!rlang::sym(logFC_column) > 1 & !!rlang::sym(FDR_column) < 0.05 ~ "Upregulated",
        !!rlang::sym(logFC_column) < -1 & !!rlang::sym(FDR_column) < 0.05 ~ "Downregulated",
        TRUE ~ "Not Selected"))
    results_table <- rbind(results_table, to_add_genes)
    ints_genes <- rbind(ints_genes, to_add_genes)
  }
  
  # Define the plot
  p <- results_table |>
    ggplot(aes(x = !!rlang::sym(logFC_column),
               y = -log10(!!rlang::sym(FDR_column)))) +
    geom_point(aes(colour = gene_type), 
               alpha = 0.8, 
               shape = 16,
               size = 1) +
    geom_hline(yintercept = y_threshold,
               linetype = "dashed") + 
    geom_vline(xintercept = c(log2(0.5), log2(2)),
               linetype = "dashed") +
    geom_point(data = dplyr::filter(ints_genes, gene_type == "up"),
               shape = 21,
               size = 2,
               fill = "tomato3", 
               colour = "black") + 
    geom_point(data = dplyr::filter(ints_genes, gene_type == "down"),
               shape = 21,
               size = 2,
               fill = "steelblue", 
               colour = "black") +
    geom_label_repel(data = ints_genes, # Add labels last to appear as the top layer  
                     aes(label = !!rlang::sym(gene_column)),
                     force = 2,
                     nudge_y = .2, nudge_x = 0.1,
                     max.overlaps = 30, size = 2) +
    scale_color_manual(values = c("Upregulated" = up_color,
                                  "Downregulated" = down_color,
                                  "Not Selected" = ns_color)) +
    scale_x_continuous(breaks = c(seq(-10, 10, 2)),     
                       limits = x_limits) +
    labs(x = x_label, y = y_label, color = "Genes") +#title = title

    #theme_void() +  # Remove background
    #theme(legend.position = "none")  # Remove legend if not needed
    theme_classic(base_size = 14) +
    theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    legend.position = "right",
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10)
  )
  
  if (!is.null(genes_of_interest)) {
    p <- p + geom_point(data = to_add_genes,
                        shape = 21,
                        size = 2.4,
                        fill = "#70e000", 
                        colour = "black")
    }
  # Save the plot
  if (!is.null(output_file)) {
    ggsave(plot = p,
           filename = output_file,
           device = format_plot,
           dpi = 300,
           units = "in",
           width = 7, height = 12)
  }
  
  return(p)
}

```

###################
#       DEG       #
###################
```{r}

# Define function to test for differential expression
test_diff_exp <- function(counts, conditions,ensembl_genes,batch=NULL,control_condition,patient=NULL) {
  
  # Argument validation
  if (!is.data.frame(counts) || !is.character(conditions) || is.null(ensembl_genes)) {
    stop("Invalid input data.")
  }
  
  
  # Ensure 'conditions' is a factor with as the reference level for this experiment, to be compared to
  conditions <- factor(conditions)
  conditions <- relevel(conditions, ref = control_condition)
  # Now defining the reference becomes the right part of the comparison where we are looking at the baseline of non.sclerotic as reference to compare each gene of the         other condition such as sclerotic now.
  
  message(paste0("Setting up DGEList"))
  d = DGEList(counts = counts, group = conditions, genes = ensembl_genes)
  
  #perform TMM normalization
  # Filter out lowly expressed genes (less than 30 counts across all samples)
  #This creates a logical vector where each entry indicates whether the corresponding gene has a CPM of at least 1 in at least 2 samples.
  keep <- rowSums(cpm(d) >= 1) >= 2# & rowSums(d$counts) >= 30
  d <- d[keep, , keep.lib.sizes = FALSE]
  
  message("Normalizing using TMM.")
  d <- calcNormFactors(d,method="TMM")
  #An MDS plot shows the similarities between samples based on their expression profiles.
  mds_plot <- plotMDS(d,col=as.numeric(as.factor(conditions)))
  mds_plot <-legend("topright", legend = levels(as.factor(conditions)), col = 1:length(levels(as.factor(conditions))), pch = 16)
  message(paste0("MDS Plot created"))
  
  
  # Save the normalized matrix
  normalized_matrix <- cpm(d, normalized.lib.sizes = TRUE)
  rownames(normalized_matrix) <- d$genes$genes
  
  
  # Convert patient or/and batch to a factor
  # Convert patient and batch to factors if they are provided
  if (!is.null(patient)) {
    patient <- factor(patient)
  }
  if (!is.null(batch)) {
    batch <- factor(batch)
    batch <- make.names(batch)
  }
  

  # Model design
  #design <- model.matrix(~0 + conditions)
  # here samples are not independet so shall not use the 0 and because of paired shall include the patient ID. without intercept is difficult to nterpret batch and patient   rather than using each of the conditions as its own group. is more typical to nclude interccept in paired and mixed#
  #    design <- model.matrix(~0 + conditions + batch + patient)

  
  #design <- model.matrix(as.formula(design_formula), data = d$samples)
  #If not using group in DGE list.
  #design <- model.matrix(~patient + conditions)
  # If declare group above in DGElist, use group instead of conditions.
  
  # Construct the design matrix formula and matrix
  if (!is.null(batch) && !is.null(patient)) {
    design_formula <- "~batch + patient + group"
    design <- model.matrix(as.formula(design_formula),data = d$samples)
  } else if (!is.null(batch)) {
    design_formula <- "~batch + group" 
    design <- model.matrix(as.formula(design_formula),data = d$samples)
  } else if (!is.null(patient)) {
    design_formula <- "~patient + group"
    design <- model.matrix(as.formula(design_formula),data = d$samples)
  } else {
    design_formula <- "~group"
    design <- model.matrix(as.formula(design_formula),data = d$samples)
  }
  
  # Output the design matrix formula
  message(paste0("Using design matrix formula: ", design_formula))
  message("Using design matrix:")
  print(head(design))
  
  message(paste0("Estimating dispersion"))
  # Fit the NB GLMs with QL methods
  d <- estimateDisp(d, design)
  #2.
  # Fit the model and perform the differential expression analysis
  library(statmod)
  message(paste0("Fitting the QLNB function."))

  fit <- glmQLFit(d, design,robust = TRUE)
  # it is kind of a shrinkage, leaving very extreme values out of the model
  message(paste0("Print dispersion plot"))
  ql_disp_plot <- plotQLDisp(fit)
  
  # Define contrasts, not in ~group
  # Get the comparison name
  non_reference_condition <- gsub(pattern = "group",replacement = "",x = colnames(design)[dim(design)[2]])
  comparison_name <- paste0(non_reference_condition,"-vs-",control_condition)
  message("Performing differential expression analysis.")
  message(paste0("Running contrast on: ",comparison_name))
  

  
  # Perform differential expression analysis
  res <- glmQLFTest(fit) #without ~0+conditions, it performs the last column as contrast, or specify column by coef=32..
  
  # Create a column with the rowSums of the samples in the comparison.
  value1 <- strsplit(x = comparison_name, split = "-vs-")[[1]][1]
  value2 <- strsplit(x = comparison_name, split = "-vs-")[[1]][2]
  
  subset_rows_v1 <- d$samples[d$samples$group == value1, , drop = FALSE]
  subset_rows_v2 <- d$samples[d$samples$group == value2, , drop = FALSE]
    
  sample_names_v1 <- rownames(subset_rows_v1)
  sample_names_v2 <- rownames(subset_rows_v2)
    
  colSum_sample_v1 <- as.matrix(rowSums(d$counts[,sample_names_v1]))
  colSum_sample_v2 <- as.matrix(rowSums(d$counts[,sample_names_v2]))
    
  colnames(colSum_sample_v1) <- paste0("colSums_",subset_rows_v1$group[1])
  colnames(colSum_sample_v2) <- paste0("colSums_",subset_rows_v2$group[1])
  
  colsums_samples <- as.data.frame(d$counts[,c(sample_names_v1,sample_names_v2)])
  colsums_samples$genes <- d$genes$genes
    
  colsums_samples$C1 <- colSum_sample_v1
  colsums_samples$C2 <- colSum_sample_v2
    
  names_to_add_after_colsums <- colnames(colsums_samples)
    
  names_to_add_after_colsums[length(names_to_add_after_colsums) -1] <- paste0("colSums_",subset_rows_v1$group[1])
  names_to_add_after_colsums[length(names_to_add_after_colsums)] <- paste0("colSums_",subset_rows_v2$group[1])
    
  colnames(colsums_samples) <- names_to_add_after_colsums
  #rownames(colsums_samples) <- d$genes$genes
  # Save
  Sys.sleep(1)
  
  # List of top genes
  top_genes <- topTags(res, n = Inf)
  top_gene_names <- top_genes$table$genes[1:10]
  # Prepare the data frame for all top genes
  samples_boxplot <- data.frame(
    samples = rep(colnames(normalized_matrix[top_gene_names,]), each=1),
    values = as.vector(t(normalized_matrix[top_gene_names,])),
    condition = rep(d$samples$group, length(top_gene_names)),
    gene = rep(top_gene_names, each = ncol(normalized_matrix))
  )

  # # Assigning colors to conditions
  # color_mapping <- c("non.sclerotic" = "red", "sclerotic" = "blue")

  # Generate the plot with subplots for each gene
  boxplot_gg <- ggplot(samples_boxplot, aes(x = condition, y = values)) +
    geom_boxplot(aes(group = condition), outlier.shape = NA) + # Remove black outlier points
    geom_jitter(width = 0.2, size = 1.5, alpha = 0.6, aes(color = condition)) + # Map color inside aes()
    #scale_color_manual(values = color_mapping) + # Map colors to conditions
    theme_minimal() +
    labs(subtitle = "Top Genes Normalized Values", x = "Condition", y = "Gene Value",) +
    facet_wrap(~ gene, scales = "free_y") # Create subplots for each gene
  
  
   # Store results and plots in a list
  results_list <- list(
    name = comparison_name,
    results = topTags(res, n = Inf),
    colsums_samples=colsums_samples,
    normalized_matrix=normalized_matrix,
    fit = fit,
    mds_plot = mds_plot,
    ql_disp_plot = ql_disp_plot,
    boxplot_gg = boxplot_gg
  )
  
  return(results_list)
  message("Differential expression analysis complete.")

}

```

# Differential gene expression.

## Separate the data in smaller groups.

```{r}
#tibia#hip#adipo#pellet#sclerotic#non_sclerotic
# Done
# tibias-> sclerotic vs nsclerotic
# Adipocytes Non-sclerotic vs Sclerotic of Tibia
# Pellet Non-sclerotic vs Sclerotic of Tibia

# hip-> sclerotic vs nsclerotic
# Adipocytes Non-sclerotic vs Sclerotic of Hip
# Pellet Non-sclerotic vs Sclerotic of Hip
```

# Save results DEG
```{r}
save_results_deg <- function(result_list, base_path) {
  # Filter results with FDR < 0.05
  filtered_results <- result_list$results$table |> dplyr::filter(FDR < 0.05)
  print(filtered_results)
  
  # Define file paths
  filtered_results_path <- paste0(base_path, "filtered_fdr_0.05_", result_list$name, ".txt")
  results_path <- paste0(base_path, result_list$name, ".txt")
  counts_path <- paste0(base_path, "counts_", result_list$name, ".csv")
  norm_counts_path <- paste0(base_path, "norm_counts_", result_list$name, ".csv")
  boxplot_topTags_path <- paste0(base_path, "boxplot_toptags_genes_", result_list$name, ".svg")
  
  # Write the tables
  write.table(filtered_results, filtered_results_path, sep = "\t", row.names = FALSE, col.names = TRUE)
  write.table(result_list$results$table, results_path, sep = "\t", row.names = FALSE, col.names = TRUE)
  write.table(result_list$colsums_samples, counts_path, sep = ",", row.names = FALSE, col.names = TRUE)
  write.table(result_list$normalized_matrix, norm_counts_path, sep = ",", row.names = FALSE, col.names = TRUE)
  
  # Save the ggplot as SVG
  svg(filename = boxplot_topTags_path, width = 12, height = 9, pointsize = 12)
  print(result_list$boxplot_gg)  # Print the ggplot object to the device
  dev.off()  # Close the device
  
  message("Results and plots have been successfully saved.")
}

# # Example usage
# base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets_patient/pellet_tibia_and_hip_"
# save_results_deg(result_list = pellets_hip_and_tibia_sclerotic_vs_non_sclerotic, base_path)




```

#########################
RUN DEA,by models using batch,patient, or nothing but the condition.
~0+patient+conditions: each group is considered independient and we measure the gene exprssion absolute difference from sample to sample or contrasts.

~patient+conditions: we use this because the data is paired, if more complicated designs to change contrasts can be done using no intercept. we meassure each gene expression using one condition as reference and how much are the difference between reference and control, assuming as above too, that there is no difference as Ho(null) and that there is difference as Ha(alternative)

by not using a intercept it is easier to make contrasts
with intercept, still can be compared many options, but it is more difficult to interpret if the diffrence is only biological and how to clearly separate it the study has more variables, despite using no intercept you can measure directly "absolute differences"

In order to set a referenc, use relevel, this is going to be set as the condition to be the intercept, to which th other condition will be compared. (both DESeq2 and edgeR)
if using for example non.sclerotic and sclerotic. and we use ~condition, we use non.sclerotic as ref, therefore the model will become, iintercept with non.sclerotic and the last column the combination of model.matrix with non.sclerotics as 0 and sclerotics as 1,defining already the contrast to be done. Where the last column is already the difference to the mean of the baseline of the non.sclerotic to sclrotic.


If we use ~0+condition each becomes a column to be tested the means of each gene.

If we use patient or batch, you need to put before the condition of nterest in the model. They will help to explain f there are differences between samples that could hide the effects of the condition.

for te logFC, the reference sample will be used as baseline to compare the experimental to the reference. such as: experiment vs reference, here the positive logFC would represent higher value of expression in experiment, if wanna change the order use relevel.
Check the boxplot_gg of the topTags that includes for this reason the baseline to the left of the plot and experiment to the right. showing graphically the normalized counts values of the topTags genes.

use edgeR manual for more and linear algebra and beatiful model fittings lessons n statistics and maths


What Happens Mathematically
Mathematically, if we denote:

𝑌𝑖𝑗

Y as the gene expression count for gene𝑖in sample 𝑗.
𝜇𝑖𝑗as the mean expression level for gene 𝑖 in sample j

𝑖ϕ i as the dispersion parameter for gene 𝑖i.

The model in edgeR can be written as:
𝑌𝑖𝑗∼Negative Binomial(𝜇𝑖𝑗,𝜙𝑖)

 where:log⁡(𝜇𝑖𝑗)=𝛽0+𝛽𝑝𝑎𝑡𝑖𝑒𝑛𝑡𝑗+𝛽𝑐𝑜𝑛𝑑𝑖𝑡𝑖𝑜𝑛𝑗log(μ ij​ )=β 0​ +β patient j +β condition j
 β 0: Intercept term (baseline log-expression level for the reference patient and condition).
 
 𝛽𝑝𝑎𝑡𝑖𝑒𝑛𝑡𝑗 : Coefficients representing the effect of patient 𝑗

𝛽𝑐𝑜𝑛𝑑𝑖𝑡𝑖𝑜𝑛𝑗: Coefficients representing the effect of the condition (sclerotic vs. non-sclerotic).

The patient coefficients (𝛽𝑝𝑎𝑡𝑖𝑒𝑛𝑡) adjust for each patient’s baseline expression level, which helps to isolate the true effect of the condition (𝛽𝑐𝑜𝑛𝑑𝑖𝑡𝑖𝑜𝑛) on gene expression. This adjustment is crucial when there is variability across patients, which could otherwise confound the condition effect.


Practical Implications
Removing Confounding Effects: By accounting for patient-specific effects, the model reduces potential confounding. For example, if some patients inherently have higher gene expression levels due to other biological factors, this variability won’t distort the estimation of the condition effect.

Improved Accuracy: Including patient in the model improves the accuracy of detecting differentially expressed genes by reducing noise from between-patient variability.

More Realistic Estimates: The inclusion of patient as a random effect or as a fixed effect (depending on how it's treated in your specific analysis) provides more realistic estimates of gene expression changes due to condition, not due to differences in baseline expression across patients.
# HIP

## HIP Adipocytes Sclerotic vs non sclerotic
```{r}
# Measure the time taken to run the function and store the result
time_taken <- system.time({
  hip_adipocytes_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_hip$matrix_raw_filtered_samples_for_DEG[,md_df_hip |>
                                                                                                                      dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                      #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                      dplyr::pull(sampleID)],
                                                             conditions = md_df_hip |>
                                                               dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                               #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                               dplyr::pull(Group),
                                                             #batch = md_df_hip |>dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>dplyr::pull(sample_number)#dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>,
                                                             patient = md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "adipocyte") |> dplyr::pull(sample_number),
                                                             ensembl_genes = rownames(m_pca_hip$matrix_raw),
                                                             control_condition = "HNSA")
  # Print the time taken to run the function
})

print(time_taken)
```
```{r}
hip_adipocytes_sclerotic_vs_non_sclerotic$name
print(hip_adipocytes_sclerotic_vs_non_sclerotic$boxplot_gg)
hip_adipocytes_sclerotic_vs_non_sclerotic$results$table
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/hip_patient/"
save_results_deg(hip_adipocytes_sclerotic_vs_non_sclerotic, base_path)
```
```{r}
volcano_hip_adipocytes_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = hip_adipocytes_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",hip_adipocytes_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```



## HIP Pellet Sclerotic vs non sclerotic
```{r}
time_taken <- system.time({
  hip_pellet_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_hip$matrix_raw_filtered_samples_for_DEG[,md_df_hip |>
                                                                                                                  dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                                  #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                  dplyr::pull(sampleID)],
                                                         conditions = md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "pellet") |> dplyr::pull(Group),
                                                         #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                         #batch = md_df_hip |>dplyr::filter(tissue_celltype_lowercase == "pellet") |>dplyr::pull(sample_number),
                                                         #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                         ensembl_genes = rownames(m_pca_hip$matrix_raw),
                                                         patient = md_df_hip |> dplyr::filter(tissue_celltype_lowercase == "pellet") |> dplyr::pull(sample_number),
                                                         control_condition = "HNSP")
  # Print the time taken to run the function
  })

print(time_taken)
```
```{r}
hip_pellet_sclerotic_vs_non_sclerotic$name
print(hip_pellet_sclerotic_vs_non_sclerotic$boxplot_gg)
hip_pellet_sclerotic_vs_non_sclerotic$results$table
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/hip_patient/"
save_results_deg(hip_pellet_sclerotic_vs_non_sclerotic, base_path)

```
```{r}
volcano_hip_pellet_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = hip_pellet_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",hip_pellet_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```


# Hip sclerotic vs non sclerotic

```{r}
time_taken <- system.time({
  hip_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_hip$matrix_raw_filtered_samples_for_DEG[,md_df_hip |>
                                                                                                           #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                           #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                           dplyr::pull(sampleID)],
                                                  conditions = md_df_hip |>
                                                    #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                    #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                    dplyr::pull(health_status),
                                                  batch = md_df_hip |>
                                                    #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                    #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                    dplyr::pull(sample_number),
                                                  ensembl_genes = rownames(m_pca_hip$matrix_raw),
                                                  control_condition = "non.sclerotic")
  # Print the time taken to run the function
  })

print(time_taken)
```
```{r}
hip_sclerotic_vs_non_sclerotic$name
print(hip_sclerotic_vs_non_sclerotic$boxplot_gg)
hip_sclerotic_vs_non_sclerotic$results$table
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/hip_patient/"
save_results_deg(hip_sclerotic_vs_non_sclerotic, base_path)
```
```{r}
volcano_hip_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = hip_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",hip_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```

```{r}
#Continue#
```

```{r}

```
# TIBIA
## Tibia Adipocytes Sclerotic vs non sclerotic
```{r}
time_taken <- system.time({
  tibia_adipocytes_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_tibia$matrix_raw_filtered_samples_for_DEG[,md_df_tibia |>
                                                                                                                          dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                          #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                          dplyr::pull(sampleID)],
                                                               conditions = md_df_tibia |>
                                                                 dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                 #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                 dplyr::pull(Group),
                                                               batch = md_df_tibia |>
                                                                 dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>dplyr::pull(sample_number),
                                                               #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>,
                                                               ensembl_genes = rownames(m_pca_tibia$matrix_raw),
                                                               control_condition = "TNSA")
  })

print(time_taken)
```
```{r}
tibia_adipocytes_sclerotic_vs_non_sclerotic$name
tibia_adipocytes_sclerotic_vs_non_sclerotic$results$table
print(tibia_adipocytes_sclerotic_vs_non_sclerotic$boxplot_gg)
```

```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/tibia_patient/"
save_results_deg(tibia_adipocytes_sclerotic_vs_non_sclerotic, base_path)
```
```{r}
volcano_tibia_adipocytes_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = tibia_adipocytes_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",tibia_adipocytes_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```


## TIBIA Pellet Sclerotic vs non sclerotic
```{r}
tibia_pellet_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_tibia$matrix_raw_filtered_samples_for_DEG[,md_df_tibia |>
                                                                                                                   dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                                   #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                   dplyr::pull(sampleID)],
                                                         conditions = md_df_tibia |>
                                                           dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                           #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                           dplyr::pull(Group),
                                                         batch = md_df_tibia |>
                                                           dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                           #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                           dplyr::pull(sample_number),
                                                         ensembl_genes = rownames(m_pca_tibia$matrix_raw),
                                                         control_condition = "TNSP")
```
```{r}
tibia_pellet_sclerotic_vs_non_sclerotic$name
tibia_pellet_sclerotic_vs_non_sclerotic$results$table
print(tibia_pellet_sclerotic_vs_non_sclerotic$boxplot_gg)
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/tibia_patient/"
save_results_deg(tibia_pellet_sclerotic_vs_non_sclerotic, base_path)
```


```{r}
volcano_tibia_pellet_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = tibia_pellet_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",tibia_pellet_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```
# Tibia sclerotic vs non sclerotic

```{r}

time_taken <- system.time({
  tibia_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_tibia$matrix_raw_filtered_samples_for_DEG[,md_df_tibia |>
                                                                                                               #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                               #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                               dplyr::pull(sampleID)],
                                                    conditions = md_df_tibia |>
                                                      #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                      #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                      dplyr::pull(health_status),
                                                    patient= md_df_tibia |> dplyr::pull(sample_number),
                                                    #batch = md_df_tibia |>dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                    #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>,
                                                    ensembl_genes = rownames(m_pca_tibia$matrix_raw),
                                                    control_condition = "non.sclerotic")
})
print(time_taken)
```
```{r}
tibia_sclerotic_vs_non_sclerotic$name
tibia_sclerotic_vs_non_sclerotic$results$table
print(tibia_sclerotic_vs_non_sclerotic$boxplot_gg)
```

```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/tibia_patient/"
save_results_deg(tibia_sclerotic_vs_non_sclerotic, base_path)
```


```{r}
volcano_tibia_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = tibia_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",tibia_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```

# Bone
## Hip and tibias

```{r}

```

```{r}
time_taken <- system.time({
  bone_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                                                       #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                       #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                       dplyr::pull(sampleID)],
                                     conditions = md_df_bone |>
                                       #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                       #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                       dplyr::pull(tissue_lowercase),
                                     batch = md_df_bone |>#dplyr::filter(tissue_celltype_lowercase == "pellet") |> 
                                       #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") 
                                       dplyr::pull(tissue_celltype_lowercase),
                                     #patient = md_df_bone |>dplyr::pull(tissue_lowercase),
                                     ensembl_genes = rownames(m_pca_samples_bone$matrix_raw),
                                     control_condition = "hip")
  
})
print(time_taken)

```

```{r}
bone_hip_vs_tibia$name
bone_hip_vs_tibia$results$table
bone_hip_vs_tibia$boxplot_gg
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/"
save_results_deg(bone_hip_vs_tibia, base_path)
```
# Non sclerotic bone
```{r}
time_taken <- system.time({
  bone_non_sclerotic_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                                                                     #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                                     #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                     dplyr::filter(health_status == "non.sclerotic") |> 
                                                                                                                     dplyr::pull(sampleID)],
                                                   conditions = md_df_bone |>
                                                     #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                     #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                     dplyr::filter(health_status == "non.sclerotic") |> 
                                                     dplyr::pull(tissue_lowercase),
                                                   batch = md_df_bone |>
                                                     #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                     #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                     dplyr::filter(health_status == "non.sclerotic") |> 
                                                     dplyr::pull(tissue_celltype_lowercase),
                                                   ensembl_genes = rownames(m_pca_samples_bone$matrix_raw),
                                                   control_condition = "hip")
  })

print(time_taken)
```
```{r}
bone_non_sclerotic_hip_vs_tibia$name
bone_non_sclerotic_hip_vs_tibia$results$table
print(bone_non_sclerotic_hip_vs_tibia$boxplot_gg)
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/non_sclerotic_"
save_results_deg(bone_non_sclerotic_hip_vs_tibia, base_path)
```
# Sclerotic bone 
```{r}
time_taken <- system.time({
  bone_sclerotic_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                                                                 #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                                                                 #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                 dplyr::filter(health_status == "sclerotic") |> 
                                                                                                                 dplyr::pull(sampleID)],
                                               conditions = md_df_bone |>                                               #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                 #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                 dplyr::filter(health_status == "sclerotic") |> 
                                                 dplyr::pull(tissue_lowercase),
                                               batch = md_df_bone |>
                                                 #dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                 #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                 dplyr::filter(health_status == "sclerotic") |>
                                               dplyr::pull(tissue_celltype_lowercase),
                                               ensembl_genes = rownames(m_pca_samples_bone$matrix_raw),
                                               control_condition = "hip")
  })

print(time_taken)
```
```{r}
bone_sclerotic_hip_vs_tibia$name
bone_sclerotic_hip_vs_tibia$results$table
print(bone_sclerotic_hip_vs_tibia$boxplot_gg)
```

```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/sclerotic_"
save_results_deg(bone_sclerotic_hip_vs_tibia, base_path)
```


<!-- # bone adipocytes -->

<!-- ```{r} -->
<!-- bone_adipocytes_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |> -->
<!--                                                                                                      dplyr::filter(tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                                                                      #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                                                                      dplyr::pull(sampleID)], -->
<!--                                                           conditions = md_df_bone |> -->
<!--                                                             dplyr::filter(tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(tissue_lowercase), -->
<!--                                                           batch = md_df_bone |> -->
<!--                                                             dplyr::filter(tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(sample_number), -->
<!--                                                           ensembl_genes = m_pca_samples_bone$matrix_raw[,2]) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/adipocytes_" -->
<!-- save_results_deg(bone_adipocytes_hip_vs_tibia[[1]], base_path) -->
<!-- ``` -->

<!-- # bone pellet -->
<!-- ```{r} -->
<!-- bone_pellet_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |> -->
<!--                                                                                                      dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                                                                      dplyr::pull(sampleID)], -->
<!--                                                           conditions = md_df_bone |> -->
<!--                                                             dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(tissue_lowercase), -->
<!--                                                           batch = md_df_bone |> -->
<!--                                                             dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             #dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(sample_number), -->
<!--                                                           ensembl_genes = m_pca_samples_bone$matrix_raw[,2]) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/pellet_" -->
<!-- save_results_deg(bone_pellet_hip_vs_tibia[[1]], base_path) -->
<!-- ``` -->




<!-- ```{r} -->
<!-- # Bone Tibia vs Hip  -->
<!-- # Bone Adipocyte -> tibia vs Hip -->
<!-- # Bone Pellet -> tibia vs Hip -->
<!-- # Bone Non Sclerotic -> tibia vs hip -->
<!-- # Bone Sclerotic -> tibia vs hip -->


<!-- #Done -->
<!-- ``` -->


<!-- # tibia vs  hip sclerotic pellet -->

<!-- ```{r} -->
<!-- bone_sclerotic_pellet_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |> -->
<!--                                                                                                      #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      dplyr::pull(sampleID)], -->
<!--                                                           conditions = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::pull(tissue_lowercase), -->
<!--                                                           batch = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::pull(sample_number), -->
<!--                                                           ensembl_genes = m_pca_samples_bone$matrix_raw[,2]) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/sclerotic_pellet_" -->
<!-- save_results_deg(bone_sclerotic_pellet_hip_vs_tibia[[1]], base_path) -->
<!-- ``` -->

<!-- # tibia vs  hip non.sclerotic pellet -->
<!-- ```{r} -->
<!-- bone_non_sclerotic_pellet_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |> -->
<!--                                                                                                      #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      dplyr::pull(sampleID)], -->
<!--                                                           conditions = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::pull(tissue_lowercase), -->
<!--                                                           batch = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::pull(sample_number), -->
<!--                                                           ensembl_genes = m_pca_samples_bone$matrix_raw[,2]) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/non_sclerotic_pellet_" -->
<!-- save_results_deg(bone_non_sclerotic_pellet_hip_vs_tibia[[1]], base_path) -->
<!-- ``` -->


<!-- #tibia vs  hip non.sclerotic adipocyte -->
<!-- ```{r} -->
<!-- bone_non_sclerotic_adipocyte_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |> -->
<!--                                                                                                      #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                                                                      dplyr::pull(sampleID)], -->
<!--                                                           conditions = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(tissue_lowercase), -->
<!--                                                           batch = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "non.sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(sample_number), -->
<!--                                                           ensembl_genes = m_pca_samples_bone$matrix_raw[,2]) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/non_sclerotic_adipocyte_" -->
<!-- save_results_deg(bone_non_sclerotic_adipocyte_hip_vs_tibia[[1]], base_path) -->
<!-- ``` -->
<!-- # tibia vs  hip sclerotic adipocyte -->
<!-- ```{r} -->
<!-- bone_sclerotic_adipocyte_hip_vs_tibia <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |> -->
<!--                                                                                                      #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                                                                      dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                                                                      dplyr::pull(sampleID)], -->
<!--                                                           conditions = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(tissue_lowercase), -->
<!--                                                           batch = md_df_bone |> -->
<!--                                                             #dplyr::filter(tissue_celltype_lowercase == "pellet") |> -->
<!--                                                             dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |> -->
<!--                                                             dplyr::pull(sample_number), -->
<!--                                                           ensembl_genes = m_pca_samples_bone$matrix_raw[,2]) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/bone/sclerotic_adipocyte_" -->
<!-- save_results_deg(bone_sclerotic_adipocyte_hip_vs_tibia[[1]], base_path) -->
<!-- ``` -->




# All adipocytes
# Adipocytes bone
## Non-sclerotic vs Sclerotic 

```{r}
time_taken <- system.time({
  adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic <- test_diff_exp(counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                                                                                        dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                                        #dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                                                                                        dplyr::pull(sampleID)],
                                                                      conditions = md_df_bone |>
                                                                        dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                        #dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                        dplyr::pull(health_status),
                                                                      patient = md_df_bone |>
                                                                        dplyr::filter(tissue_celltype_lowercase == "adipocyte") |>
                                                                        #dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
                                                                        dplyr::pull(sample_number),
                                                                      ensembl_genes = rownames(m_pca_samples_bone$matrix_raw),
                                                                      control_condition = "non.sclerotic")
  })
print(time_taken)
```


```{r}
adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic$name
adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic$results$table
print(adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic$boxplot_gg)
head(adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic$fit$design)
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/adipocytes_patient/adipocyte_tibia_and_hip"
save_results_deg(adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic, base_path)
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/adipocytes_patient/"
volcano_adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",adipocyte_hip_and_tibia_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```
# All pellets
# pellets bone
## Non-sclerotic vs Sclerotic

```{r}
time_taken <- system.time({
  pellets_hip_and_tibia_sclerotic_vs_non_sclerotic <- test_diff_exp(
    counts = m_pca_samples_bone$matrix_raw_filtered_samples_for_DEG[,md_df_bone |>
                                                                      dplyr::filter(tissue_celltype_lowercase == "pellet") |>
                                                                      #dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase=="adipocyte") |>
                                                                      dplyr::pull(sampleID)],
    conditions = md_df_bone |> dplyr::filter(tissue_celltype_lowercase == "pellet") |>
      #dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
      dplyr::pull(health_status),
    patient = md_df_bone |>
      dplyr::filter(tissue_celltype_lowercase == "pellet") |>
      #dplyr::filter(health_status == "sclerotic" & tissue_celltype_lowercase == "adipocyte") |>
      dplyr::pull(sample_number),
    ensembl_genes = rownames(m_pca_samples_bone$matrix_raw),
    control_condition = "non.sclerotic")
  
})
print(time_taken)
```


```{r}
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic$name
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic$results$table
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic$bo
head(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic$fit$desing)
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets_patient/pellet_tibia_and_hip_"
save_results_deg(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic, base_path)
```


```{r}
base_path <- "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets_patient/"
volcano_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic <- plot_volcano(results_table = pellets_hip_and_tibia_sclerotic_vs_non_sclerotic$results$table,
                                                                  format_plot="tiff",
                                                                  output_file = paste0(base_path,"Volcano/volcano_",pellets_hip_and_tibia_sclerotic_vs_non_sclerotic$name,".tiff"),
                                                                  gene_column = "genes",
                                                                  genes_of_interest = c("IL11","RCN2", "IL11", "CNTNAP2", "STMN2"))
```


# Perform enrichR
```{r}
#install.packages("enrichR")
library(enrichR)
```


```{r}
setEnrichrSite("Enrichr") # Human genes
```
```{r}

websiteLive <- getOption("enrichR.live")
if (websiteLive) {
    listEnrichrSites()
    setEnrichrSite("Enrichr") # Human genes   
}
if (websiteLive) dbs <- listEnrichrDbs()
#dbs <- listEnrichrDbs()
dbs
```

```{r}
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$results$table |> dplyr::filter(FDR < 0.05)

message(paste0("Working on sample: ",pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$name))

dbs <- c('GO_Cellular_Component_2017','GO_Molecular_Function_2017','GO_Biological_Process_2017','CellMarker_2024')
enriched <- enrichr(genes = c(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$results$table |> dplyr::filter(FDR < 0.05) |> 
                      #dplyr::filter(!grepl(x = genes,pattern = "^Gm\\d+")) |>
                      #dplyr::filter(!grepl(x = genes,pattern = "*Rik$")) |> 
                      dplyr::filter(!is.na(genes)) |> 
                      dplyr::filter(genes != "") |>
                      dplyr::filter(!grepl(x = genes, pattern = "ENSG")) |>
                      dplyr::pull(genes)), databases = dbs)

enriched
```


```{r}
enriched[["GO_Biological_Process_2017"]]
# plot the enrichment 3 from dbs GOBP
```


```{r}
enriched
enriched$CellMarker_2024
```


```{r}
plotEnrich(enriched[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value")
```

# check universe.
```{r}
library(ggplot2)
library(ggraph)

lapply(names(enriched), function(y) {
  x = pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$name
  tryCatch({
    message("Save the enrichment for comparison: ",
            x,
            " and enrichment: ", y)
    # Save the enrichment to a table
    write.table(x = enriched[[y]],
                file = paste0("~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Enrichr/", y,
                              "_for_comparison_",
                              x,
                              ".txt"),
                sep = "\t",
                row.names = FALSE,
                col.names = TRUE)
    message("Save the plot for comparison: ", x, " and enrichment: ", y)
    # Save the plot to a PDF
    ggsave(plot = plotEnrich(enriched[[y]],
                             showTerms = 30, numChar = 40, y = "Count",
                             orderBy = "Adjusted.P.value",
                             title = paste0("Enrichment analysis by Enrichr", y, " in: ", x)),
           filename = paste0("~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Enrichr/",
                             y,
                             "_for_comparison_",
                             x,
                             "_barplot.pdf"),
           device = "pdf",
           width = 12,
           height = 9)
    },
    error = function(e) {
      message("There was an error with enrichment ", y, " in comparison ", x, ": ", e)
      })
})

```

# Add all comparisons in a list
```{r}
ls(all.names = TRUE)
# Get all object names that match a pattern
comparison_names <- ls(pattern = "vs")
comparison_names
```
```{r}
# Retrieve the comparisons and put them into a list
comparisons_list <- mget(comparison_names)
comparisons_list
```

# Volcanos 
```{r}

# old function

library(ggplot2)
library(ggrepel)

volcano_plot_function <- function(df_cond, lg2fc_p, log2fc_n, p_adj, title_, colordown, colorup, down_e, up_e, top_rows_to_select_genes) {
  
  df_cond_1 <- df_cond |>
    mutate(Regulation = case_when(
      logFC > lg2fc_p & FDR < p_adj ~ "UP",
      logFC < log2fc_n & FDR < p_adj ~ "DOWN",
      TRUE ~ "NO"
    )) |>
    mutate(Regulation = replace_na(Regulation, "NO")) |>
    mutate(HS = ifelse(FDR <= (df_cond |> arrange(FDR) |> slice_head(n = top_rows_to_select_genes) |> pull(FDR)), "YES", "NO"))
  
  p1 <- ggplot(data = df_cond_1, aes(x = logFC, y = -log10(FDR), col = Regulation)) +
    geom_jitter(position = position_jitter(height = 0.0000001, width = 0.0000001, seed = 123456), size = 1) +
    scale_color_manual(labels = c(down_e, "NO", up_e), values = c(colordown, "black", colorup)) +
    geom_vline(xintercept = c(-1, 1), col = "red") +
    geom_hline(yintercept = -log10(0.05), col = "red") +
    labs(title = title_) +
    xlim(-10, 10) +
    ylim(0, 5) +
    theme(panel.grid = element_blank(),
          panel.background = element_rect(fill = "transparent"),
          panel.border = element_rect(fill = "transparent"),
          legend.title = element_text(size = 7),
          legend.text = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title = element_text(size = 7, face = "bold")) +
    geom_text_repel(data = df_cond_1, label = ifelse(df_cond_1$HS == "YES", as.character(df_cond_1$genes), ""),
                    color = "black", size = 2) +
    geom_point(position = position_jitter(height = 0.0000001, width = 0.0000001, seed = 123456), pch = 21, size = 1.1, color = "black",
               data = df_cond_1 |> filter(HS == "YES"), aes(x = logFC, y = -log10(FDR))) +
    geom_point(position = position_jitter(height = 0.0000001, width = 0.0000001, seed = 123456), pch = 21, size = 1.2, color = "black",
               data = df_cond_1 |> filter(HS == "YES"), aes(x = logFC, y = -log10(FDR)))
  
  return(list(plot = p1, data_volc = p1$data))
}
```


```{r,fig.height=4,fig.width=3}
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano <- volcano_plot_function(df_cond = pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$results$table,
                                                                                  lg2fc_p = 1,log2fc_n = -1,
                                                                                  p_adj = 0.05,
                                                                                  top_rows_to_select_genes = 10,
                                                                                  title_= "Pellets Hip and Tibia Non-sclerotic vs Sclerotic",
                                                                                  colordown = "#999999",
                                                                                  down_e ="sclerotic",
                                                                                  up_e ="non.sclerotic",
                                                                                  colorup = "#E69F00")
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano
```


# personalized volcano


```{r,fig.width=4,fig.height=5}
# new "function"

pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$results$table |>
  dplyr::mutate(gene_type = dplyr::case_when(
    logFC >= 1 & FDR < 0.05 ~ "up",
    logFC <= -1 & FDR < 0.05 ~ "down",
    TRUE ~ "ns"))


#max(without_wt_and_2_samples$log2FoldChange)
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |> dplyr::count(gene_type)



#Clean the predicted and cDNA RiK genes
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |>
  #dplyr::filter(!grepl(x = mgi_symbol,pattern = "^Gm\\d+")) |>
  #dplyr::filter(grepl(x = mgi_description,pattern = "predicted gene*")) |>
  #dplyr::filter(!grepl(x = mgi_symbol,pattern = "*Rik$")) |> 
  dplyr::filter(genes  != "")

write.table(x = pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod,
            file = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Volcano/pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod_filtered_and_empty_values_in_external_gene_name.txt",
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)


# #interesting gene list
# lists_gata4_without_WT_and_G4RS13_and_G4RS21_deseq_anno_counts$TRANSHET_vs_HET |> filter(padj < 0.05 & abs(log2FoldChange) > 1) |> 
#   arrange(padj) |> slice_max(order_by = log2FoldChange,n = 25,with_ties = FALSE) |> pull(mgi_symbol)
# lists_gata4_without_WT_and_G4RS13_and_G4RS21_deseq_anno_counts$TRANSHET_vs_HET |> filter(padj < 0.05 & abs(log2FoldChange) > 1) |> 
#   arrange(padj) |> slice_min(order_by = log2FoldChange,n = 25,with_ties = FALSE) |> pull(mgi_symbol)

ints_up <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |>
  dplyr::filter(FDR < 0.05 & logFC >= 1) |> 
  dplyr::arrange(FDR) |>
  # dplyr::slice_max(padj + log2FoldChange,n = 10) |> 
  dplyr::slice(1:10) |> 
  dplyr::pull(genes)

ints_down <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |>
  dplyr::filter(FDR <= 0.05 & logFC <= -1) |> 
  dplyr::arrange(FDR) |>
  dplyr::slice(1:10) |> 
  # dplyr::slice_min(padj + log2FoldChange,n = 10) |>
  dplyr::pull(genes)

ints <- c(ints_up,ints_down)

ints_genes <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |> filter(genes %in% ints) |> dplyr::distinct(genes,.keep_all = TRUE)

ints_down <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |> filter(genes %in% ints_down) |> dplyr::distinct(genes,.keep_all = TRUE)

ints_up <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |> filter(genes %in% ints_up) |> dplyr::distinct(genes,.keep_all = TRUE)


#add special genes of interest

to_add_gene <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |> filter(genes == "COL7A1") 
to_add_gene$gene_type <- "ns"

pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod <- rbind(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod,to_add_gene)
ints_genes <- rbind(ints_genes,to_add_gene)


# Add colour, size and alpha (transparency) 
cols <- c("up" = "#ffad73", "down" = "#26b3ff", "ns" = "grey") 
sizes <- c("up" = 2, "down" = 2, "ns" = 1) 
alphas <- c("up" = 1, "down" = 1, "ns" = 0.5)



p1 <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |>
  ggplot(aes(x = logFC,
             y = -log10(FDR))) +
  geom_point(aes(colour = gene_type), 
             alpha = 0.8, 
             shape = 16,
             size = 1) +
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(log2(0.5), log2(2)),
             linetype = "dashed") +
  geom_point(data = ints_up,
             shape = 21,
             size = 2,
             fill = "tomato3", 
             colour = "black") + 
  geom_point(data = ints_down,
             shape = 21,
             size = 2,
             fill = "steelblue", 
             colour = "black") +
  geom_point(data = to_add_gene,
             shape = 21,
             size = 3,
             fill = "green", 
             colour = "black") +
  geom_label_repel(data = ints_genes, # Add labels last to appear as the top layer  
                   aes(label = genes),
                   force = 2,
                   nudge_y = .2,nudge_x = 0.1,
                   max.overlaps = 30,size = 2) +
  scale_colour_manual(values = cols) + 
  scale_x_continuous(breaks = c(seq(-10, 10, 2)),     
                     limits = c(-10, 10)) 

ggsave(plot = p1,
       filename = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Volcano/pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod_filtered_and_empty_values_in_external_gene_name.tiff",
       device = "tiff",
       dpi = 77,
       units = "in",
       width = 7,height = 12)

p1


```

```{r}
# Add colour, size and alpha (transparency) 
cols <- c("up" = "#ffad73", "down" = "#26b3ff", "ns" = "grey") 
sizes <- c("up" = 2, "down" = 2, "ns" = 1) 
alphas <- c("up" = 1, "down" = 1, "ns" = 0.5)



p2 <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod |>
  ggplot(aes(x = logFC,
             y = -log10(FDR))) +
  geom_point(aes(colour = gene_type), 
             alpha = 0.8, 
             shape = 16,
             size = 1) +
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") +
  geom_vline(xintercept = c(log2(0.5), log2(2)),
             linetype = "dashed") +
  # geom_point(data = ints_up,
  #            shape = 21,
  #            size = 2,
  #            fill = "tomato3", 
  #            colour = "black") + 
  # geom_point(data = ints_down,
  #            shape = 21,
  #            size = 2,
  #            fill = "steelblue", 
  #            colour = "black") +
  # geom_point(data = to_add_gene,
  #            shape = 21,
  #            size = 3,
  #            fill = "blue", 
  #            colour = "black") +
  # geom_label_repel(data = ints_genes, # Add labels last to appear as the top layer  
  #                  aes(label = external_gene_name),
  #                  force = 2,
  #                  nudge_y = 1) +
  scale_colour_manual(values = cols) + 
  scale_x_continuous(breaks = c(seq(-10, 10, 2)),     
                     limits = c(-10, 10)) 


ggsave(plot = p2,
       filename = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Volcano/pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_volcano_mod_V2.tiff",
       device = "tiff",
       dpi = 77,
       units = "in",
       width = 7,height = 12)

p2
```
# Heatmap
```{r}
library(heatmaply)
library(dplyr)
library(RColorBrewer)

# Prepare the data
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$results$table %>%
  mutate(gene_type = case_when(
    logFC >= 1 & FDR < 0.05 ~ "up",
    logFC <= -1 & FDR < 0.05 ~ "down",
    TRUE ~ "ns"
  )) %>%
  filter(genes != "") %>%
  arrange(FDR) %>%
  slice_head(n = 500)

# Prepare the raw counts matrix
raw_counts <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap %>%
  select(starts_with("X"), genes) %>%
  left_join(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$colsums_samples, by = "genes") %>%
  select(starts_with("X")) %>%
  as.matrix()

rownames(raw_counts) <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap$genes



# Prepare the normalized counts matrix
norm_counts_df <- as.data.frame(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$normalized_matrix)
norm_counts_df$genes <- rownames(norm_counts_df)
norm_counts <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap %>%
  select(starts_with("X"), genes) %>%
  left_join(norm_counts_df, by = "genes") %>%
  select(starts_with("X")) %>%
  as.matrix()

rownames(norm_counts) <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap$genes
norm_counts <- na.omit(norm_counts)
```


# Raw counts
```{r}
# Define the color palette
my_palette <- rev(colorRampPalette(brewer.pal(11, "RdYlBu"))(256))

# Generate the heatmap
heatmaply(
  raw_counts,
  scale_fill_gradient_fun = scale_fill_gradientn(
    colours = c("#313695", "#4575B4", "#74ADD1", "#ABD9E9", "white", "#FDAE61", "#F46D43", "#D73027", "#A50026"),
    limits = c(-1, 1),
    values = scales::rescale(c(-1, -0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1))
  ),
  col_side_colors = tibble::tibble("<b>Column Names</b>" = md_df_bone %>% filter(tissue_celltype == "pellet") %>% pull(health_status)),
  col_side_palette = c("non.sclerotic" = "#E69F00", "sclerotic" = "#0673B3"),
  hide_colorbar = FALSE,
  showticklabels = TRUE,
  Rowv = TRUE,
  Colv = TRUE,
  plot_method = "ggplot",
  scale = "none",
  subplot_widths = c(0.2, 0.3),
  subplot_heights = c(0.1, 0.02, 0.88),
  key.title = "Expression Z-Score\nRaw Counts Values",
  grid_color = "white",
  grid_width = 0.71,
  fontsize_row = 7,
  hclust_method = "mcquitty",
  main = "pellets non.sclerotic vs sclerotic",
  file = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Heatmaps/50_heatmap_plotly.html"
)

```

# Norm counts
```{r}
# Define the color palette
my_palette <- rev(colorRampPalette(brewer.pal(11, "RdYlBu"))(256))

# Generate the heatmap
heatmaply(
  norm_counts,
  scale_fill_gradient_fun = scale_fill_gradientn(
    colours = c("#313695", "#4575B4", "#74ADD1", "#ABD9E9", "white", "#FDAE61", "#F46D43", "#D73027", "#A50026"),
    limits = c(-1, 1),
    values = scales::rescale(c(-1, -0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1))
  ),
  col_side_colors = tibble::tibble("<b>Column Names</b>" = md_df_bone %>% filter(tissue_celltype == "pellet") %>% pull(health_status)),
  col_side_palette = c("non.sclerotic" = "#E69F00", "sclerotic" = "#0673B3"),
  hide_colorbar = FALSE,
  showticklabels = TRUE,
  Rowv = TRUE,
  Colv = TRUE,
  plot_method = "ggplot",
  scale = "none",
  subplot_widths = c(0.2, 0.3),
  subplot_heights = c(0.1, 0.02, 0.88),
  key.title = "Expression Z-Score\nNormalized Counts Values",
  grid_color = "white",
  grid_width = 0.71,
  fontsize_row = 7,
  hclust_method = "mcquitty",
  main = "pellets non.sclerotic vs sclerotic",
  file = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Heatmaps/Norm_50_heatmap_plotly.html"
)

```
# Raw counts pheatmap
```{r}
library(ComplexHeatmap)
library(pheatmap)
library(svglite)
library(dplyr)

# Calculate Pearson correlation distance
distance.row <- as.dist(1 - cor(t(raw_counts)))
distance.col <- as.dist(1 - cor(raw_counts))

# Perform average linkage clustering
cluster.row <- hclust(distance.row, method = "average")
cluster.col <- hclust(distance.col, method = "average")

# Prepare annotation data
annotation_col <- as.data.frame(colnames(raw_counts))
colnames(annotation_col) <- "sample"
annotation_col$sample_2 <- annotation_col$sample

# Split the sample names into components
annotation_col <- annotation_col %>%
  separate(sample, into = c("ID", "tissue", "condition", "type"), sep = "_")
annotation_col <- annotation_col %>% select(-ID, -tissue, -type)

# Merge with metadata
annotation_col <- annotation_col %>%
  left_join(md_df_bone %>% filter(tissue_celltype == "pellet"), by = c("sample_2" = "sampleID"))

# Select relevant columns
annotation_col <- annotation_col %>% dplyr::select(condition, Gender,)#Plate.ID_tissue, Age, Gender)

rownames(annotation_col) <- colnames(raw_counts)




# Define color palette
my_palette <- colorRampPalette(rev(brewer.pal(n = 58, name = "RdYlBu")))(256)

# Define annotation colors
annotation_colors <- list(
  condition = c("non.sclerotic" = "#E69F00", "sclerotic" = "#0673B3"),
  #Plate.ID_tissue = c("plate 1_tibia" = "#1f78b4", "plate 2_tibia" = "#33a02c", "plate 3_hip" = "#e31a1c"),
  Gender = c("M" = "#a6cee3", "F" = "#fb9a99")
)

# Print annotation_colors for debugging
print(annotation_colors)

# Generate the heatmap
pheatmap_raw_counts_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic <- pheatmap(
  raw_counts,
  scale = "row",
  cluster_rows = cluster.row,
  cluster_cols = cluster.col,
  cellwidth = 12,
  cellheight = 9,
  breaks = seq(-1, 1, length.out = 101),
  annotation_col = annotation_col,
  annotation_colors = annotation_colors,
  filename = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Heatmaps/pheatmap_raw_counts_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_pearson_corr_and_linkage_100_genes.png"
)

# Display the heatmap
pheatmap_raw_counts_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic

```

# Norm counts pheatmap
```{r}
library(ComplexHeatmap)
library(pheatmap)
library(svglite)
library(dplyr)

# Calculate Pearson correlation distance
distance.row <- as.dist(1 - cor(t(norm_counts)))
distance.col <- as.dist(1 - cor(norm_counts))

# Perform average linkage clustering
cluster.row <- hclust(distance.row, method = "average")
cluster.col <- hclust(distance.col, method = "average")

# Prepare annotation data
annotation_col <- as.data.frame(colnames(norm_counts))
colnames(annotation_col) <- "sample"
annotation_col$sample_2 <- annotation_col$sample

# Split the sample names into components
annotation_col <- annotation_col %>%
  separate(sample, into = c("ID", "tissue", "condition", "type"), sep = "_")
annotation_col <- annotation_col %>% select(-ID, -tissue, -type)

# Merge with metadata
annotation_col <- annotation_col %>%
  left_join(md_df_bone %>% filter(tissue_celltype == "pellet"), by = c("sample_2" = "sampleID"))

# Select relevant columns
annotation_col <- annotation_col %>% dplyr::select(condition, Gender,)#Plate.ID_tissue, Age, Gender)

rownames(annotation_col) <- colnames(norm_counts)




# Define color palette
my_palette <- colorRampPalette(rev(brewer.pal(n = 58, name = "RdYlBu")))(256)

# Define annotation colors
annotation_colors <- list(
  condition = c("non.sclerotic" = "#E69F00", "sclerotic" = "#0673B3"),
  #Plate.ID_tissue = c("plate 1_tibia" = "#1f78b4", "plate 2_tibia" = "#33a02c", "plate 3_hip" = "#e31a1c"),
  Gender = c("M" = "#a6cee3", "F" = "#fb9a99")
)

# Print annotation_colors for debugging
print(annotation_colors)

# Generate the heatmap
pheatmap_norm_counts_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic <- pheatmap(
  norm_counts,
  scale = "row",
  cluster_rows = cluster.row,
  cluster_cols = cluster.col,
  cellwidth = 12,
  cellheight = 9,
  breaks = seq(-1, 1, length.out = 101),
  annotation_col = annotation_col,
  annotation_colors = annotation_colors,
  filename = "~/CLUSTER/Environments/2024_07_05_Lea_Lausanne_Hip_Tibia_BRBseq/Analysis_removed_samples/DGE_Results_tables/pellets/Heatmaps/pheatmap_norm_counts_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_pearson_corr_and_linkage_100_genes.png"
)

# Display the heatmap
pheatmap_norm_counts_pellets_hip_and_tibia_sclerotic_vs_non_sclerotic

```
# Model, try to predict condition (non.sclerotic vs sclerotic) using a model and the metadata.
# Randomforest first attempt
```{r,fig.height=7, fig.width=8}
library(randomForest)
library(caret)


library(heatmaply)
library(dplyr)
library(RColorBrewer)

# Prepare the data
pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$results$table |> 
  # mutate(gene_type = case_when(
  #   logFC >= 1 & FDR < 0.05 ~ "up",
  #   logFC <= -1 & FDR < 0.05 ~ "down",
  #   TRUE ~ "ns"
  # )) %>%
  # filter(genes != "") %>%
  # arrange(FDR) %>%
  # slice_head(n = 500)

# Prepare the raw counts matrix
raw_counts <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap %>%
  select(starts_with("X"), genes) %>%
  left_join(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$colsums_samples, by = "genes") %>%
  select(starts_with("X")) %>%
  as.matrix()

rownames(raw_counts) <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap$genes



# Prepare the normalized counts matrix
norm_counts_df <- as.data.frame(pellets_hip_and_tibia_sclerotic_vs_non_sclerotic[[1]]$normalized_matrix)
norm_counts_df$genes <- rownames(norm_counts_df)
norm_counts <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap %>%
  select(starts_with("X"), genes) %>%
  left_join(norm_counts_df, by = "genes") %>%
  select(starts_with("X")) %>%
  as.matrix()

rownames(norm_counts) <- pellets_hip_and_tibia_sclerotic_vs_non_sclerotic_heatmap$genes
norm_counts <- na.omit(norm_counts)





# Prepare annotation data
annotation_col <- as.data.frame(colnames(norm_counts))
colnames(annotation_col) <- "sample"
annotation_col$sample_2 <- annotation_col$sample

# Split the sample names into components
annotation_col <- annotation_col %>%
  separate(sample, into = c("ID", "tissue", "condition", "type"), sep = "_")
annotation_col <- annotation_col %>% select(-ID, -tissue, -type)

# Merge with metadata
annotation_col <- annotation_col %>%
  left_join(md_df_bone %>% filter(tissue_celltype == "pellet"), by = c("sample_2" = "sampleID"))

# Select relevant columns
annotation_col <- annotation_col %>% dplyr::select(condition, Gender
, )#Plate.ID_tissue, Age, Gender)barcodePlate.ID, X260.230.value..Nanodrop.,Concentration..ng.µL

rownames(annotation_col) <- colnames(norm_counts)




combined_data <- cbind(t(raw_counts), annotation_col)

# NO 
# NA VALUES NOT ALLOWED


# Alternatively, you can remove rows with missing values
combined_data <- na.omit(combined_data)
#column_to_remove <- "MAMDC2-AS1"
# combined_data <- combined_data %>% select(-all_of(column_to_remove))
# Replace all hyphens with dots in column names
colnames(combined_data) <- gsub("-", ".", colnames(combined_data))
# combined_data <- combined_data %>% select(-matches("AS1$"))
# combined_data <- combined_data %>% select(-matches("AS2$"))
# combined_data <- combined_data %>% select(-matches("IT1$"))
# combined_data <- combined_data %>% select(-matches("EGLN2$"))
# combined_data <- combined_data %>% select(-matches("-3$"))





# Ensure the condition column is a factor
combined_data$condition <- as.factor(combined_data$condition)

# Split the data into training and testing sets
set.seed(123) # For reproducibility
trainIndex <- createDataPartition(combined_data$condition, p = .7, 
                                  list = FALSE, 
                                  times = 1)
train_data <- combined_data[trainIndex, ]
test_data <- combined_data[-trainIndex, ]

# Train a Random Forest model
rf_model <- randomForest(condition ~ ., data = train_data, importance = TRUE)

# Predict on the test data
predictions <- predict(rf_model, test_data)

# Evaluate the model
confusionMatrix(predictions, test_data$condition)

# Feature importance
importance(rf_model)
varImpPlot(rf_model)
```
Gini: how goodd is the feature to splti the trees...
MeanDecrease: how important is the feature that if removed the model performs worst, if so the feature is important...

```{r}
# Plot variable importance
varImpPlot(rf_model, main = "Variable Importance")

```

```{r}
# Extract importance
importance <- importance(rf_model)
importance_df <- data.frame(Feature = rownames(importance), Importance = importance[, "MeanDecreaseGini"])

# Plot top 20 important features
top_features <- importance_df %>% top_n(20, Importance)
ggplot(top_features, aes(x = reorder(Feature, Importance), y = Importance)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  labs(title = "Top 20 Important Features", x = "Features", y = "Importance")

```
```{r}
library(caret)
rf_pred_class <- predict(rf_model, test_data)
confusionMatrix <- confusionMatrix(rf_pred_class, test_data$condition)
fourfoldplot(confusionMatrix$table, color = c("#CC6666", "#99CC99"), conf.level = 0, margin = 1, main = "Confusion Matrix")

```
```{r}
#Explore with more models...
```

